From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stained Glass OS <ke7oxh@gmail.com>
Date: Tue, 13 Oct 2026 23:32:51 -0700
Subject: [PATCH] vbscript: IActiveScript threads, interrupt, clone, type
 libraries, scriptlets, IDispatchEx members, text attributes (3209)

Implements the vbscript gap items: IActiveScriptError::GetSourceLineText;
IActiveScript::AddTypeLib (constants of enumerations and modules are global
names; persistent libraries survive a reset), GetCurrentScriptThreadID,
GetScriptThreadID, GetScriptThreadState, InterruptScriptThread (an abort is
not resumed by On Error Resume Next, SCRIPTINTERRUPT_RAISEEXCEPTION raises a
normal error), Clone, SetScriptState(DISCONNECTED); AddScriptlet (procedure
named after the event, bound through IBindEventHandler);
IActiveScriptDebug text attributes; IDispatchEx of class instances
(GetTypeInfo with function, property and field descriptions, GetIDsOfNames,
GetDispID flags, DeleteMember*, GetMemberProperties/Name, GetNextDispID) and
of the script scope (function invoke flags, constants, array elements,
DeleteMember*, GetMemberProperties/Name, GetNextDispID); Is on non-objects
(424), assignment to constants and procedures (501), calling a non-array
variable, a variable as a statement, Err = value, GetObject(path, class),
For Each over arrays of records, SubMatches/Match2/MatchCollection2 type
info and enumerators.

Gate fastvbb1 (64 and 32 bit), 36 mutants SG_MUTANT_VBB1_* each shown to
fail it. vbscript conformance tests 64 and 32 bit: 0 failures.

diff --git a/dlls/vbscript/global.c b/dlls/vbscript/global.c
index 2a40e16..1206e15 100644
--- a/dlls/vbscript/global.c
+++ b/dlls/vbscript/global.c
@@ -2755,8 +2755,14 @@ static HRESULT Global_CreateObject(BuiltinDisp *This, VARIANT *arg, unsigned arg
     return S_OK;
 }
 
+static BOOL is_omitted_arg(const VARIANT *v)
+{
+    return V_VT(v) == VT_ERROR && V_ERROR(v) == DISP_E_PARAMNOTFOUND;
+}
+
 static HRESULT Global_GetObject(BuiltinDisp *This, VARIANT *args, unsigned args_cnt, VARIANT *res)
 {
+    BSTR path = NULL, class = NULL;
     IBindCtx *bind_ctx;
     IUnknown *obj_unk;
     IDispatch *disp;
@@ -2766,38 +2772,101 @@ static HRESULT Global_GetObject(BuiltinDisp *This, VARIANT *args, unsigned args_
 
     TRACE("%s %s\n", args_cnt ? debugstr_variant(args) : "", args_cnt > 1 ? debugstr_variant(args+1) : "");
 
-    if(args_cnt != 1 || V_VT(args) != VT_BSTR) {
-        FIXME("unsupported args\n");
-        return E_NOTIMPL;
+    /* GetObject([pathname] [, class]) */
+    if(args_cnt && !is_omitted_arg(args)) {
+        hres = to_string(args, &path);
+        if(FAILED(hres))
+            return hres;
+    }
+    if(args_cnt > 1 && !is_omitted_arg(args+1)) {
+        hres = to_string(args+1, &class);
+        if(FAILED(hres)) {
+            SysFreeString(path);
+            return hres;
+        }
+    }
+    if(path && !*path) {
+        SysFreeString(path);
+        path = NULL;
+    }
+    if(class && !*class) {
+        SysFreeString(class);
+        class = NULL;
+    }
+
+    if(!path && !class) {
+        /* the pathname may only be left out when the class is given (the error is a guess) */
+        return MAKE_VBSERROR(VBSE_ILLEGAL_FUNC_CALL);
     }
 
     if(This->ctx->safeopt & (INTERFACE_USES_SECURITY_MANAGER|INTERFACESAFE_FOR_UNTRUSTED_DATA)) {
         WARN("blocked in current safety mode\n");
-        return VB_E_CANNOT_CREATE_OBJ;
+        hres = VB_E_CANNOT_CREATE_OBJ;
+        goto done;
     }
 
-    hres = CreateBindCtx(0, &bind_ctx);
-    if(FAILED(hres))
-        return hres;
+    obj_unk = NULL;
+    if(!path) {
+        /* a running instance of the class */
+        GUID guid;
 
-    hres = MkParseDisplayName(bind_ctx, V_BSTR(args), &eaten, &mon);
-    if(SUCCEEDED(hres)) {
-        hres = IMoniker_BindToObject(mon, bind_ctx, NULL, &IID_IUnknown, (void**)&obj_unk);
-        IMoniker_Release(mon);
+        hres = CLSIDFromProgID(class, &guid);
+        if(SUCCEEDED(hres))
+            hres = GetActiveObject(&guid, NULL, &obj_unk);
+        if(FAILED(hres)) {
+            hres = VB_E_CANNOT_CREATE_OBJ;
+            goto done;
+        }
+    }else if(class
+#ifdef SG_MUTANT_VBB1_GETOBJ
+             && 0
+#endif
+            ) {
+        /* a new object of the class that loads the file */
+        IPersistFile *persist;
+
+        obj_unk = create_object(This->ctx, class);
+        if(!obj_unk) {
+            hres = VB_E_CANNOT_CREATE_OBJ;
+            goto done;
+        }
+
+        hres = IUnknown_QueryInterface(obj_unk, &IID_IPersistFile, (void**)&persist);
+        if(SUCCEEDED(hres)) {
+            hres = IPersistFile_Load(persist, path, STGM_READ);
+            IPersistFile_Release(persist);
+        }else {
+            hres = MAKE_VBSERROR(VBSE_OLE_FILE_NOT_FOUND); /* guessed */
+        }
+        if(FAILED(hres)) {
+            IUnknown_Release(obj_unk);
+            goto done;
+        }
     }else {
-        hres = MK_E_SYNTAX;
-    }
-    IBindCtx_Release(bind_ctx);
-    if(FAILED(hres))
-        return hres;
+        hres = CreateBindCtx(0, &bind_ctx);
+        if(FAILED(hres))
+            goto done;
 
-    hres = set_object_site(This->ctx, obj_unk);
-    if(FAILED(hres)) {
-        IUnknown_Release(obj_unk);
-        return hres;
+        hres = MkParseDisplayName(bind_ctx, path, &eaten, &mon);
+        if(SUCCEEDED(hres)) {
+            hres = IMoniker_BindToObject(mon, bind_ctx, NULL, &IID_IUnknown, (void**)&obj_unk);
+            IMoniker_Release(mon);
+        }else {
+            hres = MK_E_SYNTAX;
+        }
+        IBindCtx_Release(bind_ctx);
+        if(FAILED(hres))
+            goto done;
+
+        hres = set_object_site(This->ctx, obj_unk);
+        if(FAILED(hres)) {
+            IUnknown_Release(obj_unk);
+            goto done;
+        }
     }
 
     hres = IUnknown_QueryInterface(obj_unk, &IID_IDispatch, (void**)&disp);
+    IUnknown_Release(obj_unk);
     if(SUCCEEDED(hres)) {
         if(res) {
             V_VT(res) = VT_DISPATCH;
@@ -2806,9 +2875,12 @@ static HRESULT Global_GetObject(BuiltinDisp *This, VARIANT *args, unsigned args_
             IDispatch_Release(disp);
         }
     }else {
-        FIXME("object does not support IDispatch\n");
+        WARN("object does not support IDispatch\n");
     }
 
+done:
+    SysFreeString(path);
+    SysFreeString(class);
     return hres;
 }
 
@@ -4129,6 +4201,7 @@ typedef struct {
     LONG ref;
     script_ctx_t *ctx;
     function_t *func;
+    BOOL lenient;   /* an event handler of a scriptlet: the arguments of the event are ignored */
     struct list entry;
 } FuncRef;
 
@@ -4203,7 +4276,8 @@ static HRESULT WINAPI FuncRef_Invoke(IDispatch *iface, DISPID id, REFIID riid, L
     if (!This->ctx || !This->func) return E_UNEXPECTED;
     if (!dp) dp = &empty;
     cnt = dp->cArgs - dp->cNamedArgs;
-    if (cnt != This->func->arg_cnt) return MAKE_VBSERROR(VBSE_FUNC_ARITY_MISMATCH);
+    if (This->lenient && !This->func->arg_cnt) dp = &empty;
+    else if (cnt != This->func->arg_cnt) return MAKE_VBSERROR(VBSE_FUNC_ARITY_MISMATCH);
     if (res) V_VT(res) = VT_EMPTY;
     /* called from outside a running script (an event): errors are reported */
     return exec_script(This->ctx, This->ctx->exec_top == NULL, This->func, NULL, dp, res);
@@ -4231,11 +4305,26 @@ void detach_func_refs(script_ctx_t *ctx)
     }
 }
 
+HRESULT create_func_ref(script_ctx_t *ctx, function_t *func, BOOL lenient, IDispatch **ret)
+{
+    FuncRef *ref;
+
+    if (!(ref = calloc(1, sizeof(*ref)))) return E_OUTOFMEMORY;
+    ref->IDispatch_iface.lpVtbl = &FuncRefVtbl;
+    ref->ref = 1;
+    ref->ctx = ctx;
+    ref->func = func;
+    ref->lenient = lenient;
+    list_add_tail(&ctx->func_refs, &ref->entry);
+    *ret = &ref->IDispatch_iface;
+    return S_OK;
+}
+
 static HRESULT Global_GetRef(BuiltinDisp *This, VARIANT *arg, unsigned args_cnt, VARIANT *res)
 {
     ScriptDisp *obj = This->ctx->script_obj;
     function_t *func = NULL;
-    FuncRef *ref;
+    IDispatch *disp;
     HRESULT hres;
     size_t i;
     BSTR name;
@@ -4258,30 +4347,30 @@ static HRESULT Global_GetRef(BuiltinDisp *This, VARIANT *arg, unsigned args_cnt,
 #endif
     if (!func) return MAKE_VBSERROR(VBSE_ILLEGAL_FUNC_CALL);
 
-    if (!(ref = calloc(1, sizeof(*ref)))) return E_OUTOFMEMORY;
-    ref->IDispatch_iface.lpVtbl = &FuncRefVtbl;
-    ref->ref = 1;
-    ref->ctx = This->ctx;
-    ref->func = func;
-    list_add_tail(&This->ctx->func_refs, &ref->entry);
+    if (FAILED(hres = create_func_ref(This->ctx, func, FALSE, &disp))) return hres;
 
     if (!res)
     {
-        IDispatch_Release(&ref->IDispatch_iface);
+        IDispatch_Release(disp);
         return S_OK;
     }
     V_VT(res) = VT_DISPATCH;
-    V_DISPATCH(res) = &ref->IDispatch_iface;
+    V_DISPATCH(res) = disp;
     return S_OK;
 }
 
+static HRESULT Err_Number(BuiltinDisp *This, VARIANT *args, unsigned args_cnt, VARIANT *res);
+
 static HRESULT Global_Err(BuiltinDisp *This, VARIANT *arg, unsigned args_cnt, VARIANT *res)
 {
     TRACE("\n");
 
     if(args_cnt) {
-        FIXME("Setter not supported\n");
+        /* Err = value sets the default property of the Err object, Number */
+#ifdef SG_MUTANT_VBB1_ERRSET
         return E_NOTIMPL;
+#endif
+        return Err_Number(This->ctx->err_obj, arg, args_cnt, res);
     }
 
     V_VT(res) = VT_DISPATCH;
diff --git a/dlls/vbscript/interp.c b/dlls/vbscript/interp.c
index 2c8e286..2c242c4 100644
--- a/dlls/vbscript/interp.c
+++ b/dlls/vbscript/interp.c
@@ -222,6 +222,16 @@ static HRESULT lookup_identifier(exec_ctx_t *ctx, BSTR name, vbdisp_invoke_type_
         return S_OK;
     }
 
+    /* the constants of the type libraries added to the script */
+    {
+        VARIANT *cv = lookup_typelib_const(ctx->script, name);
+        if(cv) {
+            ref->type = REF_CONST;
+            ref->u.v = cv;
+            return S_OK;
+        }
+    }
+
     item = lookup_named_item(ctx->script, name, SCRIPTITEM_ISVISIBLE);
     if(item && item->disp) {
         ref->type = REF_OBJ;
@@ -599,8 +609,13 @@ static HRESULT variant_call(exec_ctx_t *ctx, VARIANT *v, unsigned arg_cnt, VARIA
         hres = disp_call(ctx->script, V_DISPATCH(v), DISPID_VALUE, &dp, res);
         break;
     default:
-        FIXME("unsupported on %s\n", debugstr_variant(v));
+        /* a value that is neither an array nor an object cannot be called with arguments
+         * (the error is a guess: native VBScript reports a type mismatch) */
+        WARN("unsupported on %s\n", debugstr_variant(v));
+#ifdef SG_MUTANT_VBB1_CALLTYPE
         return E_NOTIMPL;
+#endif
+        return MAKE_VBSERROR(VBSE_TYPE_MISMATCH);
     }
 
     if(array) {
@@ -641,8 +656,29 @@ static HRESULT do_icall(exec_ctx_t *ctx, VARIANT *res, BSTR identifier, unsigned
             return variant_call(ctx, ref.u.v, arg_cnt, res);
 
         if(!res) {
-            FIXME("REF_VAR no res\n");
-            return E_NOTIMPL;
+            /* a variable named as a statement: an object calls its default member, anything else is no call */
+            VARIANT *v = ref.u.v;
+            VARIANT dummy;
+
+            if(V_VT(v) == (VT_VARIANT|VT_BYREF))
+                v = V_VARIANTREF(v);
+            if(V_VT(v) != VT_DISPATCH)
+                return MAKE_VBSERROR(VBSE_TYPE_MISMATCH); /* guessed */
+            if(!V_DISPATCH(v))
+                return MAKE_VBSERROR(VBSE_OBJECT_VARIABLE_NOT_SET);
+
+            dp.rgvarg = NULL;
+            dp.rgdispidNamedArgs = NULL;
+            dp.cArgs = dp.cNamedArgs = 0;
+            V_VT(&dummy) = VT_EMPTY;
+#ifdef SG_MUTANT_VBB1_STMTOBJ
+            break;
+#endif
+            hres = disp_call(ctx->script, V_DISPATCH(v), DISPID_VALUE, &dp, &dummy);
+            VariantClear(&dummy);
+            if(FAILED(hres))
+                return hres;
+            break;
         }
 
         V_VT(res) = VT_BYREF|VT_VARIANT;
@@ -910,11 +946,12 @@ static HRESULT assign_ident(exec_ctx_t *ctx, BSTR name, WORD flags, DISPPARAMS *
         hres = disp_propput(ctx->script, ref.u.d.disp, ref.u.d.id, flags, dp);
         break;
     case REF_FUNC:
-        FIXME("functions not implemented\n");
-        return E_NOTIMPL;
     case REF_CONST:
-        FIXME("REF_CONST\n");
+        /* neither a procedure name nor a constant can be assigned to */
+#ifdef SG_MUTANT_VBB1_ASSIGNCONST
         return E_NOTIMPL;
+#endif
+        return MAKE_VBSERROR(VBSE_ILLEGAL_ASSIGNMENT);
     case REF_OBJ:
         /* A host named item (like WScript or WSH) is not a variable: assigning to its name declares a
          * script variable of that name, which then hides the item. */
@@ -2202,22 +2239,23 @@ static HRESULT interp_is(exec_ctx_t *ctx)
 
     TRACE("\n");
 
+    /* both operands are popped even when the first is no object */
     stack_pop_deref(ctx, &v);
     if(V_VT(v.v) != VT_DISPATCH && V_VT(v.v) != VT_UNKNOWN) {
-        FIXME("Unhandled type %s\n", debugstr_variant(v.v));
-        hres = E_NOTIMPL;
+        WARN("Unhandled type %s\n", debugstr_variant(v.v));
+#ifndef SG_MUTANT_VBB1_ISOBJ
+        hres = MAKE_VBSERROR(VBSE_OBJECT_REQUIRED);
+#endif
     }else if(V_UNKNOWN(v.v)) {
         hres = IUnknown_QueryInterface(V_UNKNOWN(v.v), &IID_IUnknown, (void**)&r);
     }
     if(v.owned) VariantClear(v.v);
-    if(FAILED(hres))
-        return hres;
 
     stack_pop_deref(ctx, &v);
     if(V_VT(v.v) != VT_DISPATCH && V_VT(v.v) != VT_UNKNOWN) {
-        FIXME("Unhandled type %s\n", debugstr_variant(v.v));
-        hres = E_NOTIMPL;
-    }else if(V_UNKNOWN(v.v)) {
+        WARN("Unhandled type %s\n", debugstr_variant(v.v));
+        hres = MAKE_VBSERROR(VBSE_OBJECT_REQUIRED);
+    }else if(SUCCEEDED(hres) && V_UNKNOWN(v.v)) {
         hres = IUnknown_QueryInterface(V_UNKNOWN(v.v), &IID_IUnknown, (void**)&l);
     }
     if(v.owned) VariantClear(v.v);
@@ -2667,11 +2705,19 @@ static HRESULT exec_script_scoped(script_ctx_t *ctx, BOOL extern_caller, functio
     exec.func = func;
     exec.scope = scope;
     exec.prev_top = ctx->exec_top;
+    if(!exec.prev_top) {
+        /* an interrupt only applies to scripts that are running; drop stale requests */
+        ctx->interrupt_pending = 0;
+        ctx->aborting = FALSE;
+    }
     ctx->exec_top = &exec;
 
     while(exec.instr) {
         op = exec.instr->op;
-        hres = op_funcs[op](&exec);
+        if(ctx->interrupt_pending)
+            hres = take_interrupt(ctx);
+        else
+            hres = op_funcs[op](&exec);
         if(FAILED(hres)) {
             if(hres != SCRIPT_E_RECORDED) {
                 /* SCRIPT_E_RECORDED means ctx->ei is already populated */
@@ -2682,7 +2728,7 @@ static HRESULT exec_script_scoped(script_ctx_t *ctx, BOOL extern_caller, functio
             if(!ctx->ei.bstrDescription)
                 map_vbs_exception(&ctx->ei);
 
-            if(exec.resume_next) {
+            if(exec.resume_next && !ctx->aborting) {
                 unsigned stack_off;
 
                 WARN("Failed %08lx in resume next mode\n", ctx->ei.scode);
@@ -2731,6 +2777,8 @@ static HRESULT exec_script_scoped(script_ctx_t *ctx, BOOL extern_caller, functio
 
     assert(!exec.top);
     ctx->exec_top = exec.prev_top;
+    if(!exec.prev_top)
+        ctx->aborting = FALSE;   /* the aborted script has ended */
 
     if(extern_caller) {
         if(FAILED(hres)) {
diff --git a/dlls/vbscript/utils.c b/dlls/vbscript/utils.c
index d40d1b0..b83c429 100644
--- a/dlls/vbscript/utils.c
+++ b/dlls/vbscript/utils.c
@@ -96,6 +96,33 @@ static HRESULT safearray_iter_get(safearray_iter *This, VARIANT *ret)
     if(This->vt == VT_VARIANT)
         return VariantCopy(ret, (VARIANT*)elem);
 
+    if(This->vt == VT_RECORD) {
+        /* an element is a copy of the record, made with the array's record info */
+        IRecordInfo *info;
+        void *copy;
+        HRESULT hres;
+
+        hres = SafeArrayGetRecordInfo(This->sa, &info);
+        if(FAILED(hres))
+            return hres;
+        if(!info)
+            return E_INVALIDARG;
+        if(!(copy = IRecordInfo_RecordCreate(info))) {
+            IRecordInfo_Release(info);
+            return E_OUTOFMEMORY;
+        }
+        hres = IRecordInfo_RecordCopy(info, elem, copy);
+        if(FAILED(hres)) {
+            IRecordInfo_RecordDestroy(info, copy);
+            IRecordInfo_Release(info);
+            return hres;
+        }
+        V_VT(ret) = VT_RECORD;
+        V_RECORD(ret) = copy;
+        V_RECORDINFO(ret) = info;
+        return S_OK;
+    }
+
     V_VT(&ref) = This->vt | VT_BYREF;
     V_BYREF(&ref) = elem;
     return VariantCopyInd(ret, &ref);
@@ -224,9 +251,10 @@ HRESULT create_safearray_iter(SAFEARRAY *sa, BOOL owned, IEnumVARIANT **ev)
 
     if(sa && !(sa->fFeatures & FADF_VARIANT)) {
         hres = SafeArrayGetVartype(sa, &vt);
-        if(FAILED(hres) || vt == VT_RECORD || vt == VT_EMPTY) {
-            FIXME("enumeration not supported: %x\n", sa->fFeatures);
-            return E_NOTIMPL;
+        if(FAILED(hres) || vt == VT_EMPTY) {
+            /* the type of the elements is unknown (the error is a guess) */
+            WARN("element type unknown: %x\n", sa->fFeatures);
+            return MAKE_VBSERROR(VBSE_TYPE_MISMATCH);
         }
     }
 
diff --git a/dlls/vbscript/vbdisp.c b/dlls/vbscript/vbdisp.c
index e04fad8..6962237 100644
--- a/dlls/vbscript/vbdisp.c
+++ b/dlls/vbscript/vbdisp.c
@@ -119,6 +119,7 @@ static HRESULT invoke_variant_prop(script_ctx_t *ctx, VARIANT *v, WORD flags, DI
     HRESULT hres;
 
     switch(flags) {
+    case DISPATCH_METHOD: /* a variable called as a method is read (a guess: nothing else can be meant) */
     case DISPATCH_PROPERTYGET|DISPATCH_METHOD:
     case DISPATCH_PROPERTYGET:
         if(dp->cArgs) {
@@ -128,13 +129,15 @@ static HRESULT invoke_variant_prop(script_ctx_t *ctx, VARIANT *v, WORD flags, DI
                 return DISP_E_MEMBERNOTFOUND; /* That's what tests show */
             }
 
-            if (FAILED(hres = array_access(V_ARRAY(v), dp, &v)))
+            if (FAILED(hres = array_access(V_ISBYREF(v) ? *V_ARRAYREF(v) : V_ARRAY(v), dp, &v)))
             {
                 WARN("failed to access array element\n");
                 return hres;
             }
         }
 
+        if(!res)
+            return S_OK;
         hres = VariantCopyInd(res, v);
         break;
 
@@ -149,23 +152,43 @@ static HRESULT invoke_variant_prop(script_ctx_t *ctx, VARIANT *v, WORD flags, DI
             return hres;
 
         if(arg_cnt(dp)) {
-            FIXME("Arguments not supported\n");
+            /* an element of an array variable */
+            VARIANT *elem;
+
+            if(!V_ISARRAY(v)) {
+                WARN("called with arguments for non-array property\n");
+                if(own_val)
+                    VariantClear(&put_val);
+                return DISP_E_MEMBERNOTFOUND;
+            }
+#ifdef SG_MUTANT_VBB1_SDARR
             return E_NOTIMPL;
+#endif
+            hres = array_access(V_ISBYREF(v) ? *V_ARRAYREF(v) : V_ARRAY(v), dp, &elem);
+            if(FAILED(hres)) {
+                if(own_val)
+                    VariantClear(&put_val);
+                return hres;
+            }
+            v = elem;
         }
 
         if(res)
             V_VT(res) = VT_EMPTY;
 
-        if(own_val)
+        if(own_val) {
+            VariantClear(v);
             *v = put_val;
-        else
+        }else {
             hres = VariantCopyInd(v, &put_val);
+        }
         break;
     }
 
     default:
-        FIXME("unimplemented flags %x\n", flags);
-        return E_NOTIMPL;
+        /* the other combinations of invoke flags do not describe a variable */
+        WARN("unsupported flags %x\n", flags);
+        return DISP_E_MEMBERNOTFOUND;
     }
 
     return hres;
@@ -358,12 +381,24 @@ static HRESULT WINAPI DispatchEx_GetTypeInfoCount(IDispatchEx *iface, UINT *pcti
     return S_OK;
 }
 
+static HRESULT create_class_typeinfo(vbdisp_t *obj, ITypeInfo **ret);
+
 static HRESULT WINAPI DispatchEx_GetTypeInfo(IDispatchEx *iface, UINT iTInfo, LCID lcid,
                                               ITypeInfo **ppTInfo)
 {
     vbdisp_t *This = impl_from_IDispatchEx(iface);
-    FIXME("(%p)->(%u %lu %p)\n", This, iTInfo, lcid, ppTInfo);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%u %lu %p)\n", This, iTInfo, lcid, ppTInfo);
+
+    if(!ppTInfo)
+        return E_POINTER;
+    *ppTInfo = NULL;
+    if(!This->desc)
+        return E_UNEXPECTED;
+    if(iTInfo)
+        return DISP_E_BADINDEX;
+
+    return create_class_typeinfo(This, ppTInfo);
 }
 
 static HRESULT WINAPI DispatchEx_GetIDsOfNames(IDispatchEx *iface, REFIID riid,
@@ -371,9 +406,35 @@ static HRESULT WINAPI DispatchEx_GetIDsOfNames(IDispatchEx *iface, REFIID riid,
                                                 DISPID *rgDispId)
 {
     vbdisp_t *This = impl_from_IDispatchEx(iface);
-    FIXME("(%p)->(%s %p %u %lu %p)\n", This, debugstr_guid(riid), rgszNames, cNames,
-          lcid, rgDispId);
+    HRESULT hres;
+    UINT i;
+
+    TRACE("(%p)->(%s %p %u %lu %p)\n", This, debugstr_guid(riid), rgszNames, cNames, lcid, rgDispId);
+
+    if(!cNames)
+        return S_OK;
+    if(!rgszNames || !rgDispId)
+        return E_POINTER;
+    if(!This->desc)
+        return E_UNEXPECTED;
+#ifdef SG_MUTANT_VBB1_CLGIDS
     return E_NOTIMPL;
+#endif
+
+    hres = vbdisp_get_id(This, rgszNames[0], VBDISP_ANY, FALSE, rgDispId);
+    if(FAILED(hres)) {
+        for(i = 1; i < cNames; i++)
+            rgDispId[i] = DISPID_UNKNOWN;
+        return hres;
+    }
+
+    /* the names of parameters have no DISPIDs */
+    hres = S_OK;
+    for(i = 1; i < cNames; i++) {
+        rgDispId[i] = DISPID_UNKNOWN;
+        hres = DISP_E_UNKNOWNNAME;
+    }
+    return hres;
 }
 
 static HRESULT WINAPI DispatchEx_Invoke(IDispatchEx *iface, DISPID dispIdMember,
@@ -399,12 +460,13 @@ static HRESULT WINAPI DispatchEx_GetDispID(IDispatchEx *iface, BSTR bstrName, DW
     if(!This->desc)
         return E_UNEXPECTED;
 
-    /* Tests show that fdexNameCaseSensitive is ignored */
-
-    if(grfdex & ~(fdexNameEnsure|fdexNameCaseInsensitive|fdexNameCaseSensitive)) {
-        FIXME("unsupported flags %lx\n", grfdex);
-        return E_NOTIMPL;
-    }
+    /* Tests show that fdexNameCaseSensitive is ignored. An object of a script class has no dynamic
+     * members, so fdexNameEnsure cannot create any and fdexNameNoDynamicProperties changes nothing;
+     * fdexNameImplicit and fdexNameInternal only describe the lookup. */
+    grfdex &= ~(fdexNameEnsure|fdexNameCaseInsensitive|fdexNameCaseSensitive|fdexNameImplicit|fdexNameInternal|
+                fdexNameNoDynamicProperties);
+    if(grfdex)
+        WARN("unknown flags %lx\n", grfdex);
 
     return vbdisp_get_id(This, bstrName, VBDISP_ANY, FALSE, pid);
 }
@@ -437,39 +499,158 @@ static HRESULT WINAPI DispatchEx_InvokeEx(IDispatchEx *iface, DISPID id, LCID lc
     return hres;
 }
 
+/* a member id of a class instance: functions come first (id 0 is the default member), then fields */
+static BOOL vbdisp_member_public(const vbdisp_t *This, DISPID id)
+{
+    const class_desc_t *desc = This->desc;
+
+    if(id < 0)
+        return FALSE;
+    if((unsigned)id < desc->func_cnt)
+        return desc->funcs[id].name && desc->funcs[id].is_public;
+    id -= desc->func_cnt;
+    return (unsigned)id < desc->prop_cnt && desc->props[id].is_public;
+}
+
+static const WCHAR *vbdisp_member_name(const vbdisp_t *This, DISPID id)
+{
+    if((unsigned)id < This->desc->func_cnt)
+        return This->desc->funcs[id].name;
+    return This->desc->props[id - This->desc->func_cnt].name;
+}
+
 static HRESULT WINAPI DispatchEx_DeleteMemberByName(IDispatchEx *iface, BSTR bstrName, DWORD grfdex)
 {
     vbdisp_t *This = impl_from_IDispatchEx(iface);
-    FIXME("(%p)->(%s %lx)\n", This, debugstr_w(bstrName), grfdex);
-    return E_NOTIMPL;
+    DISPID id;
+
+    TRACE("(%p)->(%s %lx)\n", This, debugstr_w(bstrName), grfdex);
+
+    if(!This->desc)
+        return E_UNEXPECTED;
+
+    /* the members of a class are fixed: an unknown name is nothing to delete, a known one cannot be deleted */
+    if(FAILED(vbdisp_get_id(This, bstrName, VBDISP_ANY, FALSE, &id)))
+        return S_OK;
+#ifdef SG_MUTANT_VBB1_CLDEL
+    return S_OK;
+#endif
+    return S_FALSE;
 }
 
 static HRESULT WINAPI DispatchEx_DeleteMemberByDispID(IDispatchEx *iface, DISPID id)
 {
     vbdisp_t *This = impl_from_IDispatchEx(iface);
-    FIXME("(%p)->(%lx)\n", This, id);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%lx)\n", This, id);
+
+    if(!This->desc)
+        return E_UNEXPECTED;
+    if(!vbdisp_member_public(This, id))
+        return DISP_E_MEMBERNOTFOUND;
+    return S_FALSE;
 }
 
 static HRESULT WINAPI DispatchEx_GetMemberProperties(IDispatchEx *iface, DISPID id, DWORD grfdexFetch, DWORD *pgrfdex)
 {
     vbdisp_t *This = impl_from_IDispatchEx(iface);
-    FIXME("(%p)->(%lx %lx %p)\n", This, id, grfdexFetch, pgrfdex);
-    return E_NOTIMPL;
+    BOOL can_get, can_put, can_putref, can_call;
+
+    TRACE("(%p)->(%lx %lx %p)\n", This, id, grfdexFetch, pgrfdex);
+
+    if(!pgrfdex)
+        return E_POINTER;
+    *pgrfdex = 0;
+    if(!This->desc)
+        return E_UNEXPECTED;
+    if(!vbdisp_member_public(This, id))
+        return DISP_E_UNKNOWNNAME;
+    if(grfdexFetch & ~grfdexPropAll)
+        return E_INVALIDARG;
+
+    if((unsigned)id < This->desc->func_cnt) {
+        const vbdisp_funcprop_desc_t *f = This->desc->funcs + id;
+
+        can_get = f->entries[VBDISP_CALLGET] && f->entries[VBDISP_CALLGET]->type == FUNC_PROPGET;
+        can_call = f->entries[VBDISP_CALLGET] && f->entries[VBDISP_CALLGET]->type != FUNC_PROPGET;
+#ifdef SG_MUTANT_VBB1_CLPROPS
+        can_put = TRUE;
+#else
+        can_put = !!f->entries[VBDISP_LET];
+#endif
+        can_putref = !!f->entries[VBDISP_SET];
+        /* a function is reached with DISPATCH_PROPERTYGET too */
+        can_get |= can_call;
+    }else {
+        can_get = can_put = can_putref = TRUE;
+        can_call = FALSE;
+    }
+
+    if(grfdexFetch & (fdexPropCanGet | fdexPropCannotGet))
+        *pgrfdex |= can_get ? fdexPropCanGet : fdexPropCannotGet;
+    if(grfdexFetch & (fdexPropCanPut | fdexPropCannotPut))
+        *pgrfdex |= can_put ? fdexPropCanPut : fdexPropCannotPut;
+    if(grfdexFetch & (fdexPropCanPutRef | fdexPropCannotPutRef))
+        *pgrfdex |= can_putref ? fdexPropCanPutRef : fdexPropCannotPutRef;
+    if(grfdexFetch & (fdexPropCanCall | fdexPropCannotCall))
+        *pgrfdex |= can_call ? fdexPropCanCall : fdexPropCannotCall;
+    if(grfdexFetch & (fdexPropCanConstruct | fdexPropCannotConstruct))
+        *pgrfdex |= fdexPropCannotConstruct;
+    if(grfdexFetch & (fdexPropCanSourceEvents | fdexPropCannotSourceEvents))
+        *pgrfdex |= fdexPropCannotSourceEvents;
+    return S_OK;
 }
 
 static HRESULT WINAPI DispatchEx_GetMemberName(IDispatchEx *iface, DISPID id, BSTR *pbstrName)
 {
     vbdisp_t *This = impl_from_IDispatchEx(iface);
-    FIXME("(%p)->(%lx %p)\n", This, id, pbstrName);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%lx %p)\n", This, id, pbstrName);
+
+    if(!pbstrName)
+        return E_POINTER;
+    *pbstrName = NULL;
+    if(!This->desc)
+        return E_UNEXPECTED;
+    if(!vbdisp_member_public(This, id))
+        return DISP_E_UNKNOWNNAME;
+
+    *pbstrName = SysAllocString(vbdisp_member_name(This, id));
+    return *pbstrName ? S_OK : E_OUTOFMEMORY;
 }
 
 static HRESULT WINAPI DispatchEx_GetNextDispID(IDispatchEx *iface, DWORD grfdex, DISPID id, DISPID *pid)
 {
     vbdisp_t *This = impl_from_IDispatchEx(iface);
-    FIXME("(%p)->(%lx %lx %p)\n", This, grfdex, id, pid);
-    return E_NOTIMPL;
+    unsigned total, next;
+
+    TRACE("(%p)->(%lx %lx %p)\n", This, grfdex, id, pid);
+
+    if(!pid)
+        return E_POINTER;
+    *pid = DISPID_STARTENUM;
+    if(!This->desc)
+        return E_UNEXPECTED;
+
+    total = This->desc->func_cnt + This->desc->prop_cnt;
+    if(id == DISPID_STARTENUM)
+        next = 0;
+    else if(id >= 0 && (unsigned)id < total)
+        next = id + 1;
+    else
+        return DISP_E_UNKNOWNNAME;
+
+    /* the public members in id order */
+    for(; next < total; next++) {
+#ifdef SG_MUTANT_VBB1_CLENUM
+        if(next >= This->desc->func_cnt) continue;
+#endif
+        if(vbdisp_member_public(This, next)) {
+            *pid = next;
+            return S_OK;
+        }
+    }
+    return S_FALSE;
 }
 
 static HRESULT WINAPI DispatchEx_GetNameSpaceParent(IDispatchEx *iface, IUnknown **ppunk)
@@ -577,9 +758,65 @@ typedef struct {
     UINT num_funcs;
     struct typeinfo_func *funcs;
 
-    ScriptDisp *disp;
+    ScriptDisp *disp;       /* type info of a script's global scope */
+    vbdisp_t *obj;          /* type info of a class instance (disp is NULL) */
+    MEMBERID *var_ids;      /* class instance: member ids of the public fields */
 } ScriptTypeInfo;
 
+static function_t *get_func_from_memid(const ScriptTypeInfo *typeinfo, MEMBERID memid);
+
+/* Finds a member by id: a function (func) or a variable (name only) */
+static BOOL typeinfo_member(const ScriptTypeInfo *This, MEMBERID memid, function_t **func, const WCHAR **name)
+{
+    UINT i;
+
+    *func = NULL;
+    *name = NULL;
+
+    if (This->obj)
+    {
+        const class_desc_t *desc = This->obj->desc;
+
+        if (memid >= 0 && (UINT)memid < desc->func_cnt)
+        {
+            for (i = 0; i < This->num_funcs; i++)
+            {
+                if (This->funcs[i].memid != memid) continue;
+                *func = This->funcs[i].func;
+                *name = desc->funcs[memid].name;
+                return TRUE;
+            }
+            return FALSE;
+        }
+        for (i = 0; i < This->num_vars; i++)
+        {
+            if (This->var_ids[i] != memid) continue;
+            *name = desc->props[memid - desc->func_cnt].name;
+            return TRUE;
+        }
+        return FALSE;
+    }
+
+    *func = get_func_from_memid(This, memid);
+    if (*func)
+    {
+        *name = (*func)->name;
+        return TRUE;
+    }
+    if (memid >= 1 && (UINT)memid <= This->num_vars)
+    {
+        *name = This->disp->global_vars[memid - 1]->name;
+        return TRUE;
+    }
+    return FALSE;
+}
+
+/* member ids below 1 are never members of a script's global scope; 0 is the default member of a class */
+static inline BOOL typeinfo_bad_memid(const ScriptTypeInfo *This, MEMBERID memid)
+{
+    return memid < 0 || (!memid && !This->obj);
+}
+
 static function_t *get_func_from_memid(const ScriptTypeInfo *typeinfo, MEMBERID memid)
 {
     UINT a = 0, b = typeinfo->num_funcs;
@@ -650,10 +887,13 @@ static ULONG WINAPI ScriptTypeInfo_Release(ITypeInfo *iface)
 
     if (!ref)
     {
-        for (i = 0; i < This->num_funcs; i++)
-            release_vbscode(This->funcs[i].func->code_ctx);
+        if (This->disp)
+            for (i = 0; i < This->num_funcs; i++)
+                release_vbscode(This->funcs[i].func->code_ctx);
 
-        IDispatchEx_Release(&This->disp->IDispatchEx_iface);
+        if (This->disp) IDispatchEx_Release(&This->disp->IDispatchEx_iface);
+        if (This->obj) IDispatchEx_Release(&This->obj->IDispatchEx_iface);
+        free(This->var_ids);
         free(This->funcs);
         free(This);
     }
@@ -723,10 +963,19 @@ static HRESULT WINAPI ScriptTypeInfo_GetFuncDesc(ITypeInfo *iface, UINT index, F
 
     desc->memid = This->funcs[index].memid;
     desc->funckind = FUNC_DISPATCH;
-    desc->invkind = INVOKE_FUNC;
+    switch (func->type)
+    {
+#ifndef SG_MUTANT_VBB1_CLTI
+    case FUNC_PROPGET: desc->invkind = INVOKE_PROPERTYGET; break;
+#endif
+    case FUNC_PROPLET: desc->invkind = INVOKE_PROPERTYPUT; break;
+    case FUNC_PROPSET: desc->invkind = INVOKE_PROPERTYPUTREF; break;
+    default: desc->invkind = INVOKE_FUNC; break;
+    }
     desc->callconv = CC_STDCALL;
     desc->cParams = func->arg_cnt;
-    desc->elemdescFunc.tdesc.vt = (func->type == FUNC_SUB) ? VT_VOID : VT_VARIANT;
+    desc->elemdescFunc.tdesc.vt = (func->type == FUNC_SUB || func->type == FUNC_PROPLET || func->type == FUNC_PROPSET)
+                                  ? VT_VOID : VT_VARIANT;
 
     if (func->arg_cnt) desc->lprgelemdescParam = (ELEMDESC*)(desc + 1);
     for (i = 0; i < func->arg_cnt; i++)
@@ -749,7 +998,7 @@ static HRESULT WINAPI ScriptTypeInfo_GetVarDesc(ITypeInfo *iface, UINT index, VA
     desc = calloc(1, sizeof(*desc));
     if (!desc) return E_OUTOFMEMORY;
 
-    desc->memid = index + 1;
+    desc->memid = This->obj ? This->var_ids[index] : index + 1;
     desc->varkind = VAR_DISPATCH;
     desc->elemdescVar.tdesc.vt = VT_VARIANT;
 
@@ -762,6 +1011,7 @@ static HRESULT WINAPI ScriptTypeInfo_GetNames(ITypeInfo *iface, MEMBERID memid,
 {
     ScriptTypeInfo *This = ScriptTypeInfo_from_ITypeInfo(iface);
     ITypeInfo *disp_typeinfo;
+    const WCHAR *name;
     function_t *func;
     HRESULT hr;
     UINT i = 0;
@@ -769,10 +1019,9 @@ static HRESULT WINAPI ScriptTypeInfo_GetNames(ITypeInfo *iface, MEMBERID memid,
     TRACE("(%p)->(%ld %p %u %p)\n", This, memid, rgBstrNames, cMaxNames, pcNames);
 
     if (!rgBstrNames || !pcNames) return E_INVALIDARG;
-    if (memid <= 0) return TYPE_E_ELEMENTNOTFOUND;
+    if (typeinfo_bad_memid(This, memid)) return TYPE_E_ELEMENTNOTFOUND;
 
-    func = get_func_from_memid(This, memid);
-    if (!func && memid > This->num_vars)
+    if (!typeinfo_member(This, memid, &func, &name))
     {
         hr = get_dispatch_typeinfo(&disp_typeinfo);
         if (FAILED(hr)) return hr;
@@ -787,8 +1036,9 @@ static HRESULT WINAPI ScriptTypeInfo_GetNames(ITypeInfo *iface, MEMBERID memid,
     {
         UINT num = min(cMaxNames, func->arg_cnt + 1);
 
-        rgBstrNames[0] = SysAllocString(func->name);
+        rgBstrNames[0] = SysAllocString(name);
         if (!rgBstrNames[0]) return E_OUTOFMEMORY;
+        i = 1;
 
         for (i = 1; i < num; i++)
         {
@@ -801,7 +1051,7 @@ static HRESULT WINAPI ScriptTypeInfo_GetNames(ITypeInfo *iface, MEMBERID memid,
     }
     else
     {
-        rgBstrNames[0] = SysAllocString(This->disp->global_vars[memid - 1]->name);
+        rgBstrNames[0] = SysAllocString(name);
         if (!rgBstrNames[0]) return E_OUTOFMEMORY;
         i++;
     }
@@ -876,6 +1126,12 @@ static HRESULT WINAPI ScriptTypeInfo_GetIDsOfNames(ITypeInfo *iface, LPOLESTR *r
 
     for (i = 0; i < This->num_vars; i++)
     {
+        if (This->obj)
+        {
+            if (wcsicmp(name, This->obj->desc->props[This->var_ids[i] - This->obj->desc->func_cnt].name)) continue;
+            pMemId[0] = This->var_ids[i];
+            return S_OK;
+        }
         if (wcsicmp(name, This->disp->global_vars[i]->name)) continue;
         pMemId[0] = i + 1;
         return S_OK;
@@ -892,6 +1148,8 @@ static HRESULT WINAPI ScriptTypeInfo_Invoke(ITypeInfo *iface, PVOID pvInstance,
         DISPPARAMS *pDispParams, VARIANT *pVarResult, EXCEPINFO *pExcepInfo, UINT *puArgErr)
 {
     ScriptTypeInfo *This = ScriptTypeInfo_from_ITypeInfo(iface);
+    function_t *member_func;
+    const WCHAR *member_name;
     ITypeInfo *disp_typeinfo;
     IDispatch *disp;
     HRESULT hr;
@@ -900,9 +1158,9 @@ static HRESULT WINAPI ScriptTypeInfo_Invoke(ITypeInfo *iface, PVOID pvInstance,
           pDispParams, pVarResult, pExcepInfo, puArgErr);
 
     if (!pvInstance) return E_INVALIDARG;
-    if (memid <= 0) return TYPE_E_ELEMENTNOTFOUND;
+    if (typeinfo_bad_memid(This, memid)) return TYPE_E_ELEMENTNOTFOUND;
 
-    if (!get_func_from_memid(This, memid) && memid > This->num_vars)
+    if (!typeinfo_member(This, memid, &member_func, &member_name))
     {
         hr = get_dispatch_typeinfo(&disp_typeinfo);
         if (FAILED(hr)) return hr;
@@ -926,6 +1184,7 @@ static HRESULT WINAPI ScriptTypeInfo_GetDocumentation(ITypeInfo *iface, MEMBERID
 {
     ScriptTypeInfo *This = ScriptTypeInfo_from_ITypeInfo(iface);
     ITypeInfo *disp_typeinfo;
+    const WCHAR *name;
     function_t *func;
     HRESULT hr;
 
@@ -937,7 +1196,7 @@ static HRESULT WINAPI ScriptTypeInfo_GetDocumentation(ITypeInfo *iface, MEMBERID
 
     if (memid == MEMBERID_NIL)
     {
-        if (pBstrName && !(*pBstrName = SysAllocString(L"VBScriptTypeInfo")))
+        if (pBstrName && !(*pBstrName = SysAllocString(This->obj ? This->obj->desc->name : L"VBScriptTypeInfo")))
             return E_OUTOFMEMORY;
         if (pBstrDocString &&
             !(*pBstrDocString = SysAllocString(L"Visual Basic Scripting Type Info")))
@@ -947,10 +1206,9 @@ static HRESULT WINAPI ScriptTypeInfo_GetDocumentation(ITypeInfo *iface, MEMBERID
         }
         return S_OK;
     }
-    if (memid <= 0) return TYPE_E_ELEMENTNOTFOUND;
+    if (typeinfo_bad_memid(This, memid)) return TYPE_E_ELEMENTNOTFOUND;
 
-    func = get_func_from_memid(This, memid);
-    if (!func && memid > This->num_vars)
+    if (!typeinfo_member(This, memid, &func, &name))
     {
         hr = get_dispatch_typeinfo(&disp_typeinfo);
         if (FAILED(hr)) return hr;
@@ -961,7 +1219,7 @@ static HRESULT WINAPI ScriptTypeInfo_GetDocumentation(ITypeInfo *iface, MEMBERID
 
     if (pBstrName)
     {
-        *pBstrName = SysAllocString(func ? func->name : This->disp->global_vars[memid - 1]->name);
+        *pBstrName = SysAllocString(name);
         if (!*pBstrName) return E_OUTOFMEMORY;
     }
     return S_OK;
@@ -971,6 +1229,8 @@ static HRESULT WINAPI ScriptTypeInfo_GetDllEntry(ITypeInfo *iface, MEMBERID memi
         BSTR *pBstrDllName, BSTR *pBstrName, WORD *pwOrdinal)
 {
     ScriptTypeInfo *This = ScriptTypeInfo_from_ITypeInfo(iface);
+    function_t *member_func;
+    const WCHAR *member_name;
     ITypeInfo *disp_typeinfo;
     HRESULT hr;
 
@@ -980,7 +1240,7 @@ static HRESULT WINAPI ScriptTypeInfo_GetDllEntry(ITypeInfo *iface, MEMBERID memi
     if (pBstrName) *pBstrName = NULL;
     if (pwOrdinal) *pwOrdinal = 0;
 
-    if (!get_func_from_memid(This, memid) && memid > This->num_vars)
+    if (!typeinfo_member(This, memid, &member_func, &member_name))
     {
         hr = get_dispatch_typeinfo(&disp_typeinfo);
         if (FAILED(hr)) return hr;
@@ -1015,6 +1275,8 @@ static HRESULT WINAPI ScriptTypeInfo_GetRefTypeInfo(ITypeInfo *iface, HREFTYPE h
 static HRESULT WINAPI ScriptTypeInfo_AddressOfMember(ITypeInfo *iface, MEMBERID memid, INVOKEKIND invKind, PVOID *ppv)
 {
     ScriptTypeInfo *This = ScriptTypeInfo_from_ITypeInfo(iface);
+    function_t *member_func;
+    const WCHAR *member_name;
     ITypeInfo *disp_typeinfo;
     HRESULT hr;
 
@@ -1023,7 +1285,7 @@ static HRESULT WINAPI ScriptTypeInfo_AddressOfMember(ITypeInfo *iface, MEMBERID
     if (!ppv) return E_INVALIDARG;
     *ppv = NULL;
 
-    if (!get_func_from_memid(This, memid) && memid > This->num_vars)
+    if (!typeinfo_member(This, memid, &member_func, &member_name))
     {
         hr = get_dispatch_typeinfo(&disp_typeinfo);
         if (FAILED(hr)) return hr;
@@ -1048,6 +1310,8 @@ static HRESULT WINAPI ScriptTypeInfo_CreateInstance(ITypeInfo *iface, IUnknown *
 static HRESULT WINAPI ScriptTypeInfo_GetMops(ITypeInfo *iface, MEMBERID memid, BSTR *pBstrMops)
 {
     ScriptTypeInfo *This = ScriptTypeInfo_from_ITypeInfo(iface);
+    function_t *member_func;
+    const WCHAR *member_name;
     ITypeInfo *disp_typeinfo;
     HRESULT hr;
 
@@ -1055,7 +1319,7 @@ static HRESULT WINAPI ScriptTypeInfo_GetMops(ITypeInfo *iface, MEMBERID memid, B
 
     if (!pBstrMops) return E_INVALIDARG;
 
-    if (!get_func_from_memid(This, memid) && memid > This->num_vars)
+    if (!typeinfo_member(This, memid, &member_func, &member_name))
     {
         hr = get_dispatch_typeinfo(&disp_typeinfo);
         if (FAILED(hr)) return hr;
@@ -1235,6 +1499,59 @@ static const ITypeCompVtbl ScriptTypeCompVtbl = {
     ScriptTypeComp_BindType
 };
 
+/* The type information of a class instance lists the public functions, property procedures and fields */
+static HRESULT create_class_typeinfo(vbdisp_t *obj, ITypeInfo **ret)
+{
+    const class_desc_t *desc = obj->desc;
+    ScriptTypeInfo *type_info;
+    unsigned i, t, funcs = 0, vars = 0;
+
+    for (i = 0; i < desc->func_cnt; i++)
+    {
+        if (!desc->funcs[i].name) continue;
+        for (t = 0; t < VBDISP_ANY; t++)
+            if (desc->funcs[i].entries[t] && desc->funcs[i].entries[t]->is_public) funcs++;
+    }
+    for (i = 0; i < desc->prop_cnt; i++)
+        if (desc->props[i].is_public) vars++;
+
+    if (!(type_info = calloc(1, sizeof(*type_info)))) return E_OUTOFMEMORY;
+    type_info->funcs = calloc(funcs + 1, sizeof(*type_info->funcs));
+    type_info->var_ids = calloc(vars + 1, sizeof(*type_info->var_ids));
+    if (!type_info->funcs || !type_info->var_ids)
+    {
+        free(type_info->funcs);
+        free(type_info->var_ids);
+        free(type_info);
+        return E_OUTOFMEMORY;
+    }
+
+    type_info->ITypeInfo_iface.lpVtbl = &ScriptTypeInfoVtbl;
+    type_info->ITypeComp_iface.lpVtbl = &ScriptTypeCompVtbl;
+    type_info->ref = 1;
+    type_info->num_funcs = funcs;
+    type_info->num_vars = vars;
+    type_info->obj = obj;
+
+    for (i = 0, funcs = 0; i < desc->func_cnt; i++)
+    {
+        if (!desc->funcs[i].name) continue;
+        for (t = 0; t < VBDISP_ANY; t++)
+        {
+            if (!desc->funcs[i].entries[t] || !desc->funcs[i].entries[t]->is_public) continue;
+            type_info->funcs[funcs].memid = i;
+            type_info->funcs[funcs].func = desc->funcs[i].entries[t];
+            funcs++;
+        }
+    }
+    for (i = 0, vars = 0; i < desc->prop_cnt; i++)
+        if (desc->props[i].is_public) type_info->var_ids[vars++] = desc->func_cnt + i;
+
+    IDispatchEx_AddRef(&obj->IDispatchEx_iface);
+    *ret = &type_info->ITypeInfo_iface;
+    return S_OK;
+}
+
 static inline ScriptDisp *ScriptDisp_from_IDispatchEx(IDispatchEx *iface)
 {
     return CONTAINING_RECORD(iface, ScriptDisp, IDispatchEx_iface);
@@ -1450,12 +1767,16 @@ static HRESULT WINAPI ScriptDisp_InvokeEx(IDispatchEx *iface, DISPID id, LCID lc
         switch (wFlags)
         {
         case DISPATCH_METHOD:
+#ifndef SG_MUTANT_VBB1_SDFUNC
+        case DISPATCH_PROPERTYGET:
+#endif
         case DISPATCH_METHOD | DISPATCH_PROPERTYGET:
             hres = exec_script(This->ctx, TRUE, This->global_funcs[id - 1], NULL, pdp, pvarRes);
             break;
         default:
-            FIXME("Unsupported flags %x\n", wFlags);
-            hres = E_NOTIMPL;
+            /* a procedure cannot be assigned to */
+            WARN("Unsupported flags %x\n", wFlags);
+            hres = DISP_E_MEMBERNOTFOUND;
         }
 
         goto done;
@@ -1466,10 +1787,13 @@ static HRESULT WINAPI ScriptDisp_InvokeEx(IDispatchEx *iface, DISPID id, LCID lc
         goto done;
     }
 
-    if (This->global_vars[id - 1]->is_const)
+    if (This->global_vars[id - 1]->is_const && (wFlags & (DISPATCH_PROPERTYPUT | DISPATCH_PROPERTYPUTREF))
+#ifdef SG_MUTANT_VBB1_SDCONST
+        && 0
+#endif
+       )
     {
-        FIXME("const not supported\n");
-        hres = E_NOTIMPL;
+        hres = MAKE_VBSERROR(VBSE_ILLEGAL_ASSIGNMENT);
         goto done;
     }
 
@@ -1482,39 +1806,144 @@ done:
     return hres;
 }
 
+/* The members of the global scope: the variables (ids 1..n), then the functions (ids with
+ * DISPID_FUNCTION_MASK set); enumeration lists the public functions only */
+static BOOL script_disp_member_valid(const ScriptDisp *This, DISPID id)
+{
+    if (id & DISPID_FUNCTION_MASK)
+    {
+        id &= ~DISPID_FUNCTION_MASK;
+        return id >= 1 && (size_t)id <= This->global_funcs_cnt;
+    }
+    return id >= 1 && (size_t)id <= This->global_vars_cnt;
+}
+
 static HRESULT WINAPI ScriptDisp_DeleteMemberByName(IDispatchEx *iface, BSTR bstrName, DWORD grfdex)
 {
     ScriptDisp *This = ScriptDisp_from_IDispatchEx(iface);
-    FIXME("(%p)->(%s %lx)\n", This, debugstr_w(bstrName), grfdex);
-    return E_NOTIMPL;
+    DISPID id;
+
+    TRACE("(%p)->(%s %lx)\n", This, debugstr_w(bstrName), grfdex);
+
+    if (!This->ctx) return E_UNEXPECTED;
+
+    /* declared variables and procedures cannot be removed */
+    if (FAILED(IDispatchEx_GetDispID(&This->IDispatchEx_iface, bstrName, 0, &id)))
+        return S_OK;
+    return S_FALSE;
 }
 
 static HRESULT WINAPI ScriptDisp_DeleteMemberByDispID(IDispatchEx *iface, DISPID id)
 {
     ScriptDisp *This = ScriptDisp_from_IDispatchEx(iface);
-    FIXME("(%p)->(%lx)\n", This, id);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%lx)\n", This, id);
+
+    if (!This->ctx) return E_UNEXPECTED;
+    if (!script_disp_member_valid(This, id)) return DISP_E_MEMBERNOTFOUND;
+    return S_FALSE;
 }
 
 static HRESULT WINAPI ScriptDisp_GetMemberProperties(IDispatchEx *iface, DISPID id, DWORD grfdexFetch, DWORD *pgrfdex)
 {
     ScriptDisp *This = ScriptDisp_from_IDispatchEx(iface);
-    FIXME("(%p)->(%lx %lx %p)\n", This, id, grfdexFetch, pgrfdex);
-    return E_NOTIMPL;
+    BOOL can_get = TRUE, can_put = TRUE, can_call = FALSE;
+
+    TRACE("(%p)->(%lx %lx %p)\n", This, id, grfdexFetch, pgrfdex);
+
+    if (!pgrfdex) return E_POINTER;
+    *pgrfdex = 0;
+    if (!This->ctx) return E_UNEXPECTED;
+    if (!script_disp_member_valid(This, id)) return DISP_E_UNKNOWNNAME;
+    if (grfdexFetch & ~grfdexPropAll) return E_INVALIDARG;
+
+    if (id & DISPID_FUNCTION_MASK)
+    {
+#ifndef SG_MUTANT_VBB1_SDPROPS
+        can_call = TRUE;
+#endif
+        can_put = FALSE;
+    }
+    else if (This->global_vars[id - 1]->is_const)
+        can_put = FALSE;
+
+    if (grfdexFetch & (fdexPropCanGet | fdexPropCannotGet))
+        *pgrfdex |= can_get ? fdexPropCanGet : fdexPropCannotGet;
+    if (grfdexFetch & (fdexPropCanPut | fdexPropCannotPut))
+        *pgrfdex |= can_put ? fdexPropCanPut : fdexPropCannotPut;
+    if (grfdexFetch & (fdexPropCanPutRef | fdexPropCannotPutRef))
+        *pgrfdex |= can_put ? fdexPropCanPutRef : fdexPropCannotPutRef;
+    if (grfdexFetch & (fdexPropCanCall | fdexPropCannotCall))
+        *pgrfdex |= can_call ? fdexPropCanCall : fdexPropCannotCall;
+    if (grfdexFetch & (fdexPropCanConstruct | fdexPropCannotConstruct))
+        *pgrfdex |= fdexPropCannotConstruct;
+    if (grfdexFetch & (fdexPropCanSourceEvents | fdexPropCannotSourceEvents))
+        *pgrfdex |= fdexPropCannotSourceEvents;
+    return S_OK;
 }
 
 static HRESULT WINAPI ScriptDisp_GetMemberName(IDispatchEx *iface, DISPID id, BSTR *pbstrName)
 {
     ScriptDisp *This = ScriptDisp_from_IDispatchEx(iface);
-    FIXME("(%p)->(%lx %p)\n", This, id, pbstrName);
-    return E_NOTIMPL;
+    const WCHAR *name;
+
+    TRACE("(%p)->(%lx %p)\n", This, id, pbstrName);
+
+    if (!pbstrName) return E_POINTER;
+    *pbstrName = NULL;
+    if (!This->ctx) return E_UNEXPECTED;
+    if (!script_disp_member_valid(This, id)) return DISP_E_UNKNOWNNAME;
+
+    if (id & DISPID_FUNCTION_MASK)
+        name = This->global_funcs[(id & ~DISPID_FUNCTION_MASK) - 1]->name;
+    else
+        name = This->global_vars[id - 1]->name;
+    *pbstrName = SysAllocString(name);
+    return *pbstrName ? S_OK : E_OUTOFMEMORY;
 }
 
 static HRESULT WINAPI ScriptDisp_GetNextDispID(IDispatchEx *iface, DWORD grfdex, DISPID id, DISPID *pid)
 {
     ScriptDisp *This = ScriptDisp_from_IDispatchEx(iface);
-    FIXME("(%p)->(%lx %lx %p)\n", This, grfdex, id, pid);
-    return E_NOTIMPL;
+    size_t i;
+
+    TRACE("(%p)->(%lx %lx %p)\n", This, grfdex, id, pid);
+
+    if (!pid) return E_POINTER;
+    *pid = DISPID_STARTENUM;
+    if (!This->ctx) return E_UNEXPECTED;
+
+    if (id == DISPID_STARTENUM)
+        i = 0;
+    else if (id & DISPID_FUNCTION_MASK)
+    {
+        if (!script_disp_member_valid(This, id))
+            return DISP_E_UNKNOWNNAME;
+        i = This->global_vars_cnt + (id & ~DISPID_FUNCTION_MASK);
+    }
+    else if (id >= 1 && (size_t)id <= This->global_vars_cnt)
+        i = id;
+    else
+        return DISP_E_UNKNOWNNAME;
+
+    /* variables first, then the public functions */
+    for (; i < This->global_vars_cnt + This->global_funcs_cnt; i++)
+    {
+#ifdef SG_MUTANT_VBB1_SDENUM
+        if (i >= This->global_vars_cnt) break;
+#endif
+        if (i < This->global_vars_cnt)
+        {
+            *pid = i + 1;
+            return S_OK;
+        }
+        if (This->global_funcs[i - This->global_vars_cnt]->is_public)
+        {
+            *pid = (i - This->global_vars_cnt + 1) | DISPID_FUNCTION_MASK;
+            return S_OK;
+        }
+    }
+    return S_FALSE;
 }
 
 static HRESULT WINAPI ScriptDisp_GetNameSpaceParent(IDispatchEx *iface, IUnknown **ppunk)
diff --git a/dlls/vbscript/vbregexp.c b/dlls/vbscript/vbregexp.c
index 4a56875..cc21ee3 100644
--- a/dlls/vbscript/vbregexp.c
+++ b/dlls/vbscript/vbregexp.c
@@ -205,8 +205,21 @@ static HRESULT WINAPI SubMatches_GetTypeInfo(ISubMatches *iface,
         UINT iTInfo, LCID lcid, ITypeInfo **ppTInfo)
 {
     SubMatches *This = impl_from_ISubMatches(iface);
-    FIXME("(%p)->(%u %lu %p)\n", This, iTInfo, lcid, ppTInfo);
+
+    TRACE("(%p)->(%u %lu %p)\n", This, iTInfo, lcid, ppTInfo);
+
+    if(!ppTInfo)
+        return E_POINTER;
+    *ppTInfo = NULL;
+    if(iTInfo)
+        return DISP_E_BADINDEX;
+
+#ifdef SG_MUTANT_VBB1_RXTI
     return E_NOTIMPL;
+#endif
+    *ppTInfo = typeinfos[SubMatches_tid];
+    ITypeInfo_AddRef(*ppTInfo);
+    return S_OK;
 }
 
 static HRESULT WINAPI SubMatches_GetIDsOfNames(ISubMatches *iface,
@@ -277,11 +290,150 @@ static HRESULT WINAPI SubMatches_get_Count(ISubMatches *iface, LONG *pCount)
     return S_OK;
 }
 
+
+/* enumerator of the sub matches of a match: the strings, Empty for a group that did not take part */
+typedef struct {
+    IEnumVARIANT IEnumVARIANT_iface;
+    LONG ref;
+    ISubMatches *sm;
+    LONG pos;
+    LONG count;
+} SubMatchesEnum;
+
+static inline SubMatchesEnum *impl_from_SubMatchesEnum(IEnumVARIANT *iface)
+{
+    return CONTAINING_RECORD(iface, SubMatchesEnum, IEnumVARIANT_iface);
+}
+
+static HRESULT WINAPI SubMatchesEnum_QueryInterface(IEnumVARIANT *iface, REFIID riid, void **ppv)
+{
+    SubMatchesEnum *This = impl_from_SubMatchesEnum(iface);
+
+    if(IsEqualGUID(riid, &IID_IUnknown) || IsEqualGUID(riid, &IID_IEnumVARIANT)) {
+        *ppv = &This->IEnumVARIANT_iface;
+        IUnknown_AddRef((IUnknown*)*ppv);
+        return S_OK;
+    }
+    *ppv = NULL;
+    return E_NOINTERFACE;
+}
+
+static ULONG WINAPI SubMatchesEnum_AddRef(IEnumVARIANT *iface)
+{
+    return InterlockedIncrement(&impl_from_SubMatchesEnum(iface)->ref);
+}
+
+static ULONG WINAPI SubMatchesEnum_Release(IEnumVARIANT *iface)
+{
+    SubMatchesEnum *This = impl_from_SubMatchesEnum(iface);
+    LONG ref = InterlockedDecrement(&This->ref);
+
+    if(!ref) {
+        ISubMatches_Release(This->sm);
+        free(This);
+    }
+    return ref;
+}
+
+static HRESULT WINAPI SubMatchesEnum_Next(IEnumVARIANT *iface, ULONG celt, VARIANT *rgVar, ULONG *pCeltFetched)
+{
+    SubMatchesEnum *This = impl_from_SubMatchesEnum(iface);
+    ULONG i;
+    HRESULT hres;
+
+    TRACE("(%p)->(%lu %p %p)\n", This, celt, rgVar, pCeltFetched);
+
+    if(!rgVar || (!pCeltFetched && celt != 1))
+        return E_POINTER;
+
+    for(i = 0; i < celt && This->pos < This->count; i++) {
+        VariantInit(rgVar + i);
+        hres = ISubMatches_get_Item(This->sm, This->pos, rgVar + i);
+        if(FAILED(hres)) {
+            while(i--)
+                VariantClear(rgVar + i);
+            if(pCeltFetched)
+                *pCeltFetched = 0;
+            return hres;
+        }
+        This->pos++;
+    }
+    if(pCeltFetched)
+        *pCeltFetched = i;
+    return i == celt ? S_OK : S_FALSE;
+}
+
+static HRESULT WINAPI SubMatchesEnum_Skip(IEnumVARIANT *iface, ULONG celt)
+{
+    SubMatchesEnum *This = impl_from_SubMatchesEnum(iface);
+    ULONG left = This->count - This->pos;
+
+    TRACE("(%p)->(%lu)\n", This, celt);
+
+    This->pos += min(celt, left);
+    return celt <= left ? S_OK : S_FALSE;
+}
+
+static HRESULT WINAPI SubMatchesEnum_Reset(IEnumVARIANT *iface)
+{
+    impl_from_SubMatchesEnum(iface)->pos = 0;
+    return S_OK;
+}
+
+static const IEnumVARIANTVtbl SubMatchesEnum_Vtbl;
+
+static HRESULT create_sub_matches_enum(ISubMatches *sm, LONG pos, IEnumVARIANT **ret)
+{
+    SubMatchesEnum *This = calloc(1, sizeof(*This));
+
+    if(!This)
+        return E_OUTOFMEMORY;
+    This->IEnumVARIANT_iface.lpVtbl = &SubMatchesEnum_Vtbl;
+    This->ref = 1;
+    This->sm = sm;
+    ISubMatches_AddRef(sm);
+    This->pos = pos;
+    ISubMatches_get_Count(sm, &This->count);
+    *ret = &This->IEnumVARIANT_iface;
+    return S_OK;
+}
+
+static HRESULT WINAPI SubMatchesEnum_Clone(IEnumVARIANT *iface, IEnumVARIANT **ppEnum)
+{
+    SubMatchesEnum *This = impl_from_SubMatchesEnum(iface);
+
+    TRACE("(%p)->(%p)\n", This, ppEnum);
+
+    if(!ppEnum)
+        return E_POINTER;
+#ifdef SG_MUTANT_VBB1_RXCLONE
+    return create_sub_matches_enum(This->sm, 0, ppEnum);
+#endif
+    return create_sub_matches_enum(This->sm, This->pos, ppEnum);
+}
+
+static const IEnumVARIANTVtbl SubMatchesEnum_Vtbl = {
+    SubMatchesEnum_QueryInterface,
+    SubMatchesEnum_AddRef,
+    SubMatchesEnum_Release,
+    SubMatchesEnum_Next,
+    SubMatchesEnum_Skip,
+    SubMatchesEnum_Reset,
+    SubMatchesEnum_Clone
+};
+
 static HRESULT WINAPI SubMatches_get__NewEnum(ISubMatches *iface, IUnknown **ppEnum)
 {
     SubMatches *This = impl_from_ISubMatches(iface);
-    FIXME("(%p)->(%p)\n", This, ppEnum);
+
+    TRACE("(%p)->(%p)\n", This, ppEnum);
+
+    if(!ppEnum)
+        return E_POINTER;
+#ifdef SG_MUTANT_VBB1_RXENUM
     return E_NOTIMPL;
+#endif
+    return create_sub_matches_enum(&This->ISubMatches_iface, 0, (IEnumVARIANT**)ppEnum);
 }
 
 static const ISubMatchesVtbl SubMatchesVtbl = {
@@ -411,8 +563,18 @@ static HRESULT WINAPI Match2_GetTypeInfo(IMatch2 *iface,
         UINT iTInfo, LCID lcid, ITypeInfo **ppTInfo)
 {
     Match2 *This = impl_from_IMatch2(iface);
-    FIXME("(%p)->(%u %lu %p)\n", This, iTInfo, lcid, ppTInfo);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%u %lu %p)\n", This, iTInfo, lcid, ppTInfo);
+
+    if(!ppTInfo)
+        return E_POINTER;
+    *ppTInfo = NULL;
+    if(iTInfo)
+        return DISP_E_BADINDEX;
+
+    *ppTInfo = typeinfos[Match2_tid];
+    ITypeInfo_AddRef(*ppTInfo);
+    return S_OK;
 }
 
 static HRESULT WINAPI Match2_GetIDsOfNames(IMatch2 *iface,
@@ -733,11 +895,20 @@ static HRESULT WINAPI MatchCollectionEnum_Reset(IEnumVARIANT *iface)
     return S_OK;
 }
 
+static HRESULT create_enum_variant_mc2(IMatchCollection2*,ULONG,IEnumVARIANT**);
+
 static HRESULT WINAPI MatchCollectionEnum_Clone(IEnumVARIANT *iface, IEnumVARIANT **ppEnum)
 {
     MatchCollectionEnum *This = impl_from_IMatchCollectionEnum(iface);
-    FIXME("(%p)->(%p)\n", This, ppEnum);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, ppEnum);
+
+    if(!ppEnum)
+        return E_POINTER;
+#ifdef SG_MUTANT_VBB1_RXCLONE
+    return create_enum_variant_mc2(This->mc, 0, ppEnum);
+#endif
+    return create_enum_variant_mc2(This->mc, This->pos, ppEnum);
 }
 
 static const IEnumVARIANTVtbl MatchCollectionEnum_Vtbl = {
@@ -849,8 +1020,18 @@ static HRESULT WINAPI MatchCollection2_GetTypeInfo(IMatchCollection2 *iface,
         UINT iTInfo, LCID lcid, ITypeInfo **ppTInfo)
 {
     MatchCollection2 *This = impl_from_IMatchCollection2(iface);
-    FIXME("(%p)->(%u %lu %p)\n", This, iTInfo, lcid, ppTInfo);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%u %lu %p)\n", This, iTInfo, lcid, ppTInfo);
+
+    if(!ppTInfo)
+        return E_POINTER;
+    *ppTInfo = NULL;
+    if(iTInfo)
+        return DISP_E_BADINDEX;
+
+    *ppTInfo = typeinfos[MatchCollection2_tid];
+    ITypeInfo_AddRef(*ppTInfo);
+    return S_OK;
 }
 
 static HRESULT WINAPI MatchCollection2_GetIDsOfNames(IMatchCollection2 *iface,
diff --git a/dlls/vbscript/vbscript.c b/dlls/vbscript/vbscript.c
index 2832557..2f2a1aa 100644
--- a/dlls/vbscript/vbscript.c
+++ b/dlls/vbscript/vbscript.c
@@ -64,6 +64,7 @@ typedef struct {
     DWORD_PTR cookie;
     unsigned line;
     unsigned character;
+    BSTR line_text;
 } VBScriptError;
 
 static inline WCHAR *heap_pool_strdup(heap_pool_t *heap, const WCHAR *str)
@@ -278,6 +279,109 @@ void release_named_item(named_item_t *item)
     free(item);
 }
 
+typedef struct {
+    struct list entry;
+    ITypeLib *lib;
+    DWORD flags;
+    GUID guid;
+    DWORD major, minor;
+} typelib_entry_t;
+
+/* Drops the type libraries of the script (the persistent ones stay unless all is TRUE) and the
+ * constants looked up in them */
+static void release_typelibs(script_ctx_t *ctx, BOOL all)
+{
+    typelib_entry_t *iter, *next;
+
+    while(ctx->typelib_vars) {
+        dynamic_var_t *var = ctx->typelib_vars;
+
+        ctx->typelib_vars = var->next;
+        release_dynamic_var(var);
+        free((WCHAR*)var->name);
+        free(var);
+    }
+
+    LIST_FOR_EACH_ENTRY_SAFE(iter, next, &ctx->typelibs, typelib_entry_t, entry) {
+#ifndef SG_MUTANT_VBB1_TLPERSIST
+        if(!all && (iter->flags & SCRIPTTYPELIB_ISPERSISTENT))
+            continue;
+#endif
+        list_remove(&iter->entry);
+        ITypeLib_Release(iter->lib);
+        free(iter);
+    }
+}
+
+/* The constants of the enumerations and modules of the added type libraries are global names of
+ * the script. A constant is read once; its value then lives as long as the libraries. */
+VARIANT *lookup_typelib_const(script_ctx_t *ctx, const WCHAR *name)
+{
+    typelib_entry_t *entry;
+    dynamic_var_t *var;
+    UINT i, count;
+
+    for(var = ctx->typelib_vars; var; var = var->next) {
+        if(!wcsicmp(var->name, name))
+            return &var->v;
+    }
+
+#ifdef SG_MUTANT_VBB1_TYPELIB
+    return NULL;
+#endif
+    LIST_FOR_EACH_ENTRY(entry, &ctx->typelibs, typelib_entry_t, entry) {
+        count = ITypeLib_GetTypeInfoCount(entry->lib);
+        for(i = 0; i < count; i++) {
+            ITypeInfo *info;
+            TYPEATTR *attr;
+            UINT v;
+
+            if(FAILED(ITypeLib_GetTypeInfo(entry->lib, i, &info)))
+                continue;
+            if(FAILED(ITypeInfo_GetTypeAttr(info, &attr))) {
+                ITypeInfo_Release(info);
+                continue;
+            }
+            if(attr->typekind == TKIND_ENUM || attr->typekind == TKIND_MODULE) {
+                for(v = 0; v < attr->cVars; v++) {
+                    VARDESC *vd;
+                    BSTR vname;
+                    UINT n;
+
+                    if(FAILED(ITypeInfo_GetVarDesc(info, v, &vd)))
+                        continue;
+                    if(vd->varkind == VAR_CONST && vd->lpvarValue &&
+                       SUCCEEDED(ITypeInfo_GetNames(info, vd->memid, &vname, 1, &n)) && n) {
+                        BOOL match = !wcsicmp(vname, name);
+
+                        SysFreeString(vname);
+                        if(match && (var = calloc(1, sizeof(*var)))) {
+                            var->name = wcsdup(name);
+                            if(var->name && SUCCEEDED(VariantCopy(&var->v, vd->lpvarValue))) {
+                                var->is_const = TRUE;
+                                var->next = ctx->typelib_vars;
+                                ctx->typelib_vars = var;
+                            }else {
+                                free((WCHAR*)var->name);
+                                free(var);
+                                var = NULL;
+                            }
+                            ITypeInfo_ReleaseVarDesc(info, vd);
+                            ITypeInfo_ReleaseTypeAttr(info, attr);
+                            ITypeInfo_Release(info);
+                            return var ? &var->v : NULL;
+                        }
+                    }
+                    ITypeInfo_ReleaseVarDesc(info, vd);
+                }
+            }
+            ITypeInfo_ReleaseTypeAttr(info, attr);
+            ITypeInfo_Release(info);
+        }
+    }
+    return NULL;
+}
+
 static void release_script(script_ctx_t *ctx)
 {
     named_item_t *item, *item_next;
@@ -286,6 +390,7 @@ static void release_script(script_ctx_t *ctx)
     collect_objects(ctx);
     detach_func_refs(ctx);
     clear_ei(&ctx->ei);
+    release_typelibs(ctx, FALSE);
 
     LIST_FOR_EACH_ENTRY_SAFE(code, code_next, &ctx->code_list, vbscode_t, entry)
     {
@@ -348,6 +453,7 @@ static void release_code_list(script_ctx_t *ctx)
 
 static void release_named_item_list(script_ctx_t *ctx)
 {
+    release_typelibs(ctx, TRUE);
     while(!list_empty(&ctx->named_items)) {
         named_item_t *iter = LIST_ENTRY(list_head(&ctx->named_items), named_item_t, entry);
         list_remove(&iter->entry);
@@ -518,8 +624,10 @@ static ULONG WINAPI VBScriptError_Release(IActiveScriptError *iface)
 
     TRACE("(%p) ref=%ld\n", This, ref);
 
-    if(!ref)
+    if(!ref) {
+        SysFreeString(This->line_text);
         free(This);
+    }
 
     return ref;
 }
@@ -555,8 +663,20 @@ static HRESULT WINAPI VBScriptError_GetSourcePosition(IActiveScriptError *iface,
 static HRESULT WINAPI VBScriptError_GetSourceLineText(IActiveScriptError *iface, BSTR *source)
 {
     VBScriptError *This = impl_from_IActiveScriptError(iface);
-    FIXME("(%p)->(%p)\n", This, source);
+
+    TRACE("(%p)->(%p)\n", This, source);
+
+    if(!source)
+        return E_POINTER;
+    *source = NULL;
+#ifdef SG_MUTANT_VBB1_LINETEXT
     return E_NOTIMPL;
+#endif
+    if(!This->line_text)
+        return E_FAIL;
+
+    *source = SysAllocString(This->line_text);
+    return *source ? S_OK : E_OUTOFMEMORY;
 }
 
 static const IActiveScriptErrorVtbl VBScriptErrorVtbl = {
@@ -579,6 +699,7 @@ HRESULT report_script_error(script_ctx_t *ctx, const vbscode_t *code, unsigned l
     error->IActiveScriptError_iface.lpVtbl = &VBScriptErrorVtbl;
 
     error->ref = 1;
+    error->line_text = NULL;
     error->ei = ctx->ei;
     memset(&ctx->ei, 0, sizeof(ctx->ei));
     result = error->ei.scode;
@@ -593,6 +714,11 @@ HRESULT report_script_error(script_ctx_t *ctx, const vbscode_t *code, unsigned l
     }
     error->character = code->source + loc - nl;
 
+    /* the line the error is in, without its line break */
+    p = nl;
+    while(*p && *p != '\n' && *p != '\r') p++;
+    error->line_text = SysAllocStringLen(nl, p - nl);
+
     hres = IActiveScriptSite_OnScriptError(ctx->site, &error->IActiveScriptError_iface);
     IActiveScriptError_Release(&error->IActiveScriptError_iface);
     return hres == S_OK ? SCRIPT_E_REPORTED : result;
@@ -655,6 +781,8 @@ static ULONG WINAPI VBScript_Release(IActiveScript *iface)
     if(!ref) {
         decrease_state(This, SCRIPTSTATE_CLOSED);
         detach_global_objects(This->ctx);
+        clear_ei(&This->ctx->interrupt_ei);
+        DeleteCriticalSection(&This->ctx->interrupt_cs);
         IServiceProvider_Release(&This->ctx->vbcaller->IServiceProvider_iface);
         free(This->ctx);
         free(This);
@@ -761,8 +889,18 @@ static HRESULT WINAPI VBScript_SetScriptState(IActiveScript *iface, SCRIPTSTATE
         decrease_state(This, SCRIPTSTATE_CLOSED);
         return S_OK;
     case SCRIPTSTATE_DISCONNECTED:
-        FIXME("unimplemented SCRIPTSTATE_DISCONNECTED\n");
+        if(This->state == SCRIPTSTATE_CLOSED)
+            return E_UNEXPECTED;
+#ifdef SG_MUTANT_VBB1_DISC
         return S_OK;
+#endif
+        if(This->state == SCRIPTSTATE_CONNECTED) {
+            /* the script keeps its state, only the event sinks are disconnected */
+            decrease_state(This, SCRIPTSTATE_DISCONNECTED);
+            return S_OK;
+        }
+        exec_queued_code(This->ctx);
+        break;
     default:
         FIXME("unimplemented state %d\n", ss);
         return E_NOTIMPL;
@@ -857,8 +995,45 @@ static HRESULT WINAPI VBScript_AddTypeLib(IActiveScript *iface, REFGUID rguidTyp
         DWORD dwMajor, DWORD dwMinor, DWORD dwFlags)
 {
     VBScript *This = impl_from_IActiveScript(iface);
-    FIXME("(%p)->(%s %ld %ld %ld)\n", This, debugstr_guid(rguidTypeLib), dwMajor, dwMinor, dwFlags);
-    return E_NOTIMPL;
+    typelib_entry_t *entry;
+    ITypeLib *lib;
+    HRESULT hres;
+
+    TRACE("(%p)->(%s %ld %ld %lx)\n", This, debugstr_guid(rguidTypeLib), dwMajor, dwMinor, dwFlags);
+
+    if(This->thread_id != GetCurrentThreadId() || This->state == SCRIPTSTATE_CLOSED)
+        return E_UNEXPECTED;
+
+#ifndef SG_MUTANT_VBB1_TLFLAGS
+    if(dwFlags & ~(SCRIPTTYPELIB_ISCONTROL|SCRIPTTYPELIB_ISPERSISTENT))
+        return E_INVALIDARG;
+#endif
+
+    /* the same library is added once */
+    LIST_FOR_EACH_ENTRY(entry, &This->ctx->typelibs, typelib_entry_t, entry) {
+        if(IsEqualGUID(&entry->guid, rguidTypeLib) && entry->major == dwMajor && entry->minor == dwMinor) {
+            entry->flags |= dwFlags;
+            return S_OK;
+        }
+    }
+
+    hres = LoadRegTypeLib(rguidTypeLib, dwMajor, dwMinor, This->ctx->lcid, &lib);
+    if(FAILED(hres)) {
+        WARN("LoadRegTypeLib failed: %08lx\n", hres);
+        return hres;
+    }
+
+    if(!(entry = malloc(sizeof(*entry)))) {
+        ITypeLib_Release(lib);
+        return E_OUTOFMEMORY;
+    }
+    entry->lib = lib;
+    entry->flags = dwFlags;
+    entry->guid = *rguidTypeLib;
+    entry->major = dwMajor;
+    entry->minor = dwMinor;
+    list_add_tail(&This->ctx->typelibs, &entry->entry);
+    return S_OK;
 }
 
 static HRESULT WINAPI VBScript_GetScriptDispatch(IActiveScript *iface, LPCOLESTR pstrItemName, IDispatch **ppdisp)
@@ -892,39 +1067,200 @@ static HRESULT WINAPI VBScript_GetCurrentScriptThreadID(IActiveScript *iface,
                                                        SCRIPTTHREADID *pstridThread)
 {
     VBScript *This = impl_from_IActiveScript(iface);
-    FIXME("(%p)->()\n", This);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, pstridThread);
+
+    if(!pstridThread)
+        return E_POINTER;
+
+    /* the engine runs on its one thread; the script thread id is that thread's Win32 id */
+    if(!This->thread_id || This->state == SCRIPTSTATE_CLOSED)
+        return E_UNEXPECTED;
+#ifdef SG_MUTANT_VBB1_TID
+    *pstridThread = 0;
+#else
+    *pstridThread = This->thread_id;
+#endif
+    return S_OK;
 }
 
 static HRESULT WINAPI VBScript_GetScriptThreadID(IActiveScript *iface,
                                                 DWORD dwWin32ThreadId, SCRIPTTHREADID *pstidThread)
 {
     VBScript *This = impl_from_IActiveScript(iface);
-    FIXME("(%p)->()\n", This);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%lx %p)\n", This, dwWin32ThreadId, pstidThread);
+
+    if(!pstidThread)
+        return E_POINTER;
+    if(!This->thread_id || This->state == SCRIPTSTATE_CLOSED)
+        return E_UNEXPECTED;
+    if(dwWin32ThreadId != This->thread_id)
+        return E_INVALIDARG; /* guessed: not a thread of this engine */
+    *pstidThread = dwWin32ThreadId;
+    return S_OK;
 }
 
 static HRESULT WINAPI VBScript_GetScriptThreadState(IActiveScript *iface,
         SCRIPTTHREADID stidThread, SCRIPTTHREADSTATE *pstsState)
 {
     VBScript *This = impl_from_IActiveScript(iface);
-    FIXME("(%p)->()\n", This);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%lx %p)\n", This, stidThread, pstsState);
+
+    if(!pstsState)
+        return E_POINTER;
+    if(!This->thread_id || This->state == SCRIPTSTATE_CLOSED)
+        return E_UNEXPECTED;
+    if(stidThread != SCRIPTTHREADID_CURRENT && stidThread != SCRIPTTHREADID_BASE && stidThread != This->thread_id)
+        return E_INVALIDARG;
+
+#ifdef SG_MUTANT_VBB1_THREADSTATE
+    *pstsState = SCRIPTTHREADSTATE_NOTINSCRIPT;
+#else
+    *pstsState = ((volatile script_ctx_t*)This->ctx)->exec_top ? SCRIPTTHREADSTATE_RUNNING : SCRIPTTHREADSTATE_NOTINSCRIPT;
+#endif
+    return S_OK;
+}
+
+/* Takes the interrupt requested by InterruptScriptThread: the error to raise in the running script
+ * is recorded in ctx->ei. Without SCRIPTINTERRUPT_RAISEEXCEPTION the script is aborted: On Error
+ * Resume Next does not resume it. */
+HRESULT take_interrupt(script_ctx_t *ctx)
+{
+    EnterCriticalSection(&ctx->interrupt_cs);
+    if(InterlockedExchange(&ctx->interrupt_pending, 0)) {
+        if(!(ctx->interrupt_flags & SCRIPTINTERRUPT_RAISEEXCEPTION))
+#ifndef SG_MUTANT_VBB1_ABORT
+            ctx->aborting = TRUE;
+#else
+            ;
+#endif
+        clear_ei(&ctx->ei);
+        ctx->ei = ctx->interrupt_ei;
+        memset(&ctx->interrupt_ei, 0, sizeof(ctx->interrupt_ei));
+    }
+    LeaveCriticalSection(&ctx->interrupt_cs);
+    return SCRIPT_E_RECORDED;
 }
 
 static HRESULT WINAPI VBScript_InterruptScriptThread(IActiveScript *iface,
         SCRIPTTHREADID stidThread, const EXCEPINFO *pexcepinfo, DWORD dwFlags)
 {
     VBScript *This = impl_from_IActiveScript(iface);
-    FIXME("(%p)->()\n", This);
-    return E_NOTIMPL;
+    script_ctx_t *ctx = This->ctx;
+
+    TRACE("(%p)->(%lx %p %lx)\n", This, stidThread, pexcepinfo, dwFlags);
+
+    if(!pexcepinfo)
+        return E_POINTER;
+    if(dwFlags & ~SCRIPTINTERRUPT_ALL_FLAGS)
+        return E_INVALIDARG;
+    if(!This->thread_id || This->state == SCRIPTSTATE_CLOSED || This->state == SCRIPTSTATE_UNINITIALIZED)
+        return E_UNEXPECTED;
+    if(stidThread != SCRIPTTHREADID_ALL && stidThread != SCRIPTTHREADID_CURRENT && stidThread != SCRIPTTHREADID_BASE
+       && stidThread != This->thread_id)
+        return E_INVALIDARG;
+
+    EnterCriticalSection(&ctx->interrupt_cs);
+    clear_ei(&ctx->interrupt_ei);
+    /* the error reported to the aborted script is the one in the EXCEPINFO */
+    if(pexcepinfo->scode)
+        ctx->interrupt_ei.scode = pexcepinfo->scode;
+    else if(pexcepinfo->wCode)
+        ctx->interrupt_ei.scode = MAKE_HRESULT(SEVERITY_ERROR, FACILITY_DISPATCH, pexcepinfo->wCode);
+    else
+        ctx->interrupt_ei.scode = E_ABORT;
+    ctx->interrupt_ei.wCode = pexcepinfo->wCode;
+    ctx->interrupt_ei.dwHelpContext = pexcepinfo->dwHelpContext;
+    ctx->interrupt_ei.bstrSource = pexcepinfo->bstrSource ? SysAllocString(pexcepinfo->bstrSource) : NULL;
+    ctx->interrupt_ei.bstrDescription = pexcepinfo->bstrDescription ? SysAllocString(pexcepinfo->bstrDescription) : NULL;
+    ctx->interrupt_ei.bstrHelpFile = pexcepinfo->bstrHelpFile ? SysAllocString(pexcepinfo->bstrHelpFile) : NULL;
+    ctx->interrupt_flags = dwFlags;
+#ifndef SG_MUTANT_VBB1_INTERRUPT
+    InterlockedExchange(&ctx->interrupt_pending, 1);
+#endif
+    LeaveCriticalSection(&ctx->interrupt_cs);
+    return S_OK;
 }
 
 static HRESULT WINAPI VBScript_Clone(IActiveScript *iface, IActiveScript **ppscript)
 {
     VBScript *This = impl_from_IActiveScript(iface);
-    FIXME("(%p)->()\n", This);
-    return E_NOTIMPL;
+    typelib_entry_t *tl;
+    named_item_t *item;
+    vbscode_t *code;
+    VBScript *clone;
+    HRESULT hres;
+
+    TRACE("(%p)->(%p)\n", This, ppscript);
+
+    if(!ppscript)
+        return E_POINTER;
+    *ppscript = NULL;
+
+    if(!This->is_initialized || !This->ctx->site || This->state == SCRIPTSTATE_CLOSED
+       || This->state == SCRIPTSTATE_UNINITIALIZED || This->thread_id != GetCurrentThreadId())
+        return E_UNEXPECTED;
+
+    hres = VBScriptFactory_CreateInstance(NULL, NULL, &IID_IActiveScript, (void**)ppscript);
+    if(FAILED(hres))
+        return hres;
+    clone = impl_from_IActiveScript(*ppscript);
+
+    /* the clone is initialized, waiting for its site */
+    clone->is_initialized = TRUE;
+    clone->ctx->safeopt = This->ctx->safeopt;
+    clone->ctx->lcid = This->ctx->lcid;
+    clone->ctx->codepage = This->ctx->codepage;
+
+    /* named items are duplicated without their objects (the new site supplies them), as are the type libraries */
+#ifndef SG_MUTANT_VBB1_CLONEITEMS
+    LIST_FOR_EACH_ENTRY(item, &This->ctx->named_items, named_item_t, entry) {
+        named_item_t *copy = calloc(1, sizeof(*copy));
+
+        if(copy && (copy->name = wcsdup(item->name))) {
+            copy->ref = 1;
+            copy->flags = item->flags;
+            list_add_tail(&clone->ctx->named_items, &copy->entry);
+        }else {
+            free(copy);
+            hres = E_OUTOFMEMORY;
+        }
+    }
+#endif
+    LIST_FOR_EACH_ENTRY(tl, &This->ctx->typelibs, typelib_entry_t, entry) {
+        typelib_entry_t *copy = malloc(sizeof(*copy));
+
+        if(!copy) {
+            hres = E_OUTOFMEMORY;
+            continue;
+        }
+        *copy = *tl;
+        ITypeLib_AddRef(copy->lib);
+        list_add_tail(&clone->ctx->typelibs, &copy->entry);
+    }
+
+    /* persistent code is what a reset engine would run again once it is started */
+#ifndef SG_MUTANT_VBB1_CLONECODE
+    LIST_FOR_EACH_ENTRY(code, &This->ctx->code_list, vbscode_t, entry) {
+        vbscode_t *copy;
+
+        if(!code->is_persistent || code->named_item)
+            continue;   /* code attached to a named item is not carried over */
+        hres = compile_script(clone->ctx, code->source, NULL, NULL, code->cookie, code->start_line,
+                              SCRIPTTEXT_ISPERSISTENT | SCRIPTTEXT_SG_NOREPORT, &copy);
+        if(FAILED(hres))
+            break;
+        copy->pending_exec = TRUE;
+    }
+#endif
+    if(FAILED(hres)) {
+        IActiveScript_Close(*ppscript);
+        IActiveScript_Release(*ppscript);
+        *ppscript = NULL;
+    }
+    return hres;
 }
 
 static const IActiveScriptVtbl VBScriptVtbl = {
@@ -969,22 +1305,190 @@ static ULONG WINAPI VBScriptDebug_Release(IActiveScriptDebug *iface)
     return IActiveScript_Release(&This->IActiveScript_iface);
 }
 
+static const WCHAR * const vbs_keywords[] = {
+    L"and", L"byref", L"byval", L"call", L"case", L"class", L"const", L"default", L"dim", L"do", L"each",
+    L"else", L"elseif", L"empty", L"end", L"eqv", L"erase", L"error", L"exit", L"explicit", L"false", L"for",
+    L"function", L"get", L"goto", L"if", L"imp", L"in", L"is", L"let", L"loop", L"me", L"mod", L"new", L"next",
+    L"not", L"nothing", L"null", L"on", L"option", L"or", L"preserve", L"private", L"property", L"public",
+    L"randomize", L"redim", L"rem", L"resume", L"select", L"set", L"step", L"stop", L"sub", L"then", L"to",
+    L"true", L"until", L"wend", L"while", L"with", L"xor"
+};
+
+static BOOL is_vbs_keyword(const WCHAR *s, size_t len)
+{
+    unsigned i;
+
+    for(i = 0; i < ARRAY_SIZE(vbs_keywords); i++) {
+        if(wcslen(vbs_keywords[i]) == len && !wcsnicmp(vbs_keywords[i], s, len))
+            return TRUE;
+    }
+    return FALSE;
+}
+
+static BOOL is_vbs_ident_char(WCHAR c)
+{
+    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '_' || c > 0x7f;
+}
+
+/* Classifies the characters of VBScript source: comments (' and Rem to the end of the line), strings,
+ * date literals (marked as strings; a guess), numbers (decimal, &H hexadecimal, &O octal), the
+ * operators and the reserved words. The first character of a Function or Sub that starts a
+ * procedure is also marked as a function start. The text after the delimiter is not source. */
+static HRESULT get_text_attributes(const WCHAR *code, ULONG len, const WCHAR *delimiter, SOURCE_TEXT_ATTR *attr)
+{
+#ifdef SG_MUTANT_VBB1_ATTRS
+    if(attr) memset(attr, 0, len * sizeof(*attr));
+    return S_OK;
+#endif
+    ULONG i = 0, j, end = len;
+    int prev_word = 0;  /* 1: the previous word was End or Exit */
+    BOOL stmt_start = TRUE;
+
+    if(!attr)
+        return E_POINTER;
+    memset(attr, 0, len * sizeof(*attr));
+    if(!code)
+        return len ? E_POINTER : S_OK;
+
+#ifndef SG_MUTANT_VBB1_ATTRDELIM
+    if(delimiter && *delimiter) {
+        size_t dlen = wcslen(delimiter);
+
+        for(j = 0; j + dlen <= len; j++) {
+            if(!wcsncmp(code + j, delimiter, dlen)) {
+                end = j;
+                break;
+            }
+        }
+        for(j = end; j < len; j++)
+            attr[j] = SOURCETEXT_ATTR_NONSOURCE;
+    }
+#endif
+
+    while(i < end) {
+        WCHAR c = code[i];
+
+        if(c == '\n' || c == '\r' || c == ':') {
+            stmt_start = TRUE;
+            prev_word = 0;
+            i++;
+            continue;
+        }
+        if(c == ' ' || c == '\t') {
+            i++;
+            continue;
+        }
+        if(c == '\'') {
+            for(j = i; j < end && code[j] != '\n' && code[j] != '\r'; j++)
+#ifdef SG_MUTANT_VBB1_ATTRCOMMENT
+                attr[j] = SOURCETEXT_ATTR_STRING;
+#else
+                attr[j] = SOURCETEXT_ATTR_COMMENT;
+#endif
+            i = j;
+            continue;
+        }
+        if(c == '"') {
+            j = i + 1;
+            while(j < end && code[j] != '\n' && code[j] != '\r') {
+                if(code[j] == '"') {
+                    if(j + 1 < end && code[j + 1] == '"') { j += 2; continue; }
+                    j++;
+                    break;
+                }
+                j++;
+            }
+            for(; i < j; i++)
+                attr[i] = SOURCETEXT_ATTR_STRING;
+            stmt_start = FALSE;
+            continue;
+        }
+        if(c == '#') {
+            for(j = i + 1; j < end && code[j] != '#' && code[j] != '\n' && code[j] != '\r'; j++);
+            if(j < end && code[j] == '#')
+                j++;
+            for(; i < j; i++)
+                attr[i] = SOURCETEXT_ATTR_STRING;
+            stmt_start = FALSE;
+            continue;
+        }
+        if(is_digit(c) || (c == '.' && i + 1 < end && is_digit(code[i + 1])) ||
+           (c == '&' && i + 1 < end && (code[i + 1] == 'h' || code[i + 1] == 'H' || code[i + 1] == 'o' || code[i + 1] == 'O')
+            && i + 2 < end && iswxdigit(code[i + 2]))) {
+            j = i;
+            if(c == '&') {
+                j += 2;
+                while(j < end && iswxdigit(code[j])) j++;
+                if(j < end && code[j] == '&') j++;
+            }else {
+                while(j < end && (is_digit(code[j]) || code[j] == '.')) j++;
+                if(j < end && (code[j] == 'e' || code[j] == 'E')) {
+                    ULONG k = j + 1;
+                    if(k < end && (code[k] == '+' || code[k] == '-')) k++;
+                    if(k < end && is_digit(code[k])) {
+                        while(k < end && is_digit(code[k])) k++;
+                        j = k;
+                    }
+                }
+            }
+            for(; i < j; i++)
+                attr[i] = SOURCETEXT_ATTR_NUMBER;
+            stmt_start = FALSE;
+            continue;
+        }
+        if(is_vbs_ident_char(c)) {
+            for(j = i; j < end && is_vbs_ident_char(code[j]); j++);
+            if(is_vbs_keyword(code + i, j - i)) {
+                size_t wl = j - i;
+                BOOL decl = !prev_word && ((wl == 8 && !wcsnicmp(code + i, L"function", 8)) || (wl == 3 && !wcsnicmp(code + i, L"sub", 3)));
+
+                if(wl == 3 && !wcsnicmp(code + i, L"rem", 3) && stmt_start && (j >= end || code[j] == ' ' || code[j] == '\t' ||
+                                                                                code[j] == '\n' || code[j] == '\r')) {
+                    for(j = i; j < end && code[j] != '\n' && code[j] != '\r'; j++)
+                        attr[j] = SOURCETEXT_ATTR_COMMENT;
+                    i = j;
+                    continue;
+                }
+                for(; i < j; i++)
+                    attr[i] = SOURCETEXT_ATTR_KEYWORD;
+                if(decl)
+                    attr[j - wl] |= SOURCETEXT_ATTR_FUNCTION_START;
+                prev_word = (wl == 3 && !wcsnicmp(code + j - wl, L"end", 3)) || (wl == 4 && !wcsnicmp(code + j - wl, L"exit", 4));
+            }else {
+                prev_word = 0;
+            }
+            i = j;
+            stmt_start = FALSE;
+            continue;
+        }
+        if(wcschr(L"+-*/\\^&=<>", c))
+            attr[i] = SOURCETEXT_ATTR_OPERATOR;
+        stmt_start = FALSE;
+        prev_word = 0;
+        i++;
+    }
+    return S_OK;
+}
+
 static HRESULT WINAPI VBScriptDebug_GetScriptTextAttributes(IActiveScriptDebug *iface,
         LPCOLESTR code, ULONG len, LPCOLESTR delimiter, DWORD flags, SOURCE_TEXT_ATTR *attr)
 {
     VBScript *This = impl_from_IActiveScriptDebug(iface);
-    FIXME("(%p)->(%s %lu %s %#lx %p)\n", This, debugstr_w(code), len,
-          debugstr_w(delimiter), flags, attr);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s %lu %s %#lx %p)\n", This, debugstr_wn(code, len), len, debugstr_w(delimiter), flags, attr);
+
+    return get_text_attributes(code, len, delimiter, attr);
 }
 
 static HRESULT WINAPI VBScriptDebug_GetScriptletTextAttributes(IActiveScriptDebug *iface,
         LPCOLESTR code, ULONG len, LPCOLESTR delimiter, DWORD flags, SOURCE_TEXT_ATTR *attr)
 {
     VBScript *This = impl_from_IActiveScriptDebug(iface);
-    FIXME("(%p)->(%s %lu %s %#lx %p)\n", This, debugstr_w(code), len,
-          debugstr_w(delimiter), flags, attr);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s %lu %s %#lx %p)\n", This, debugstr_wn(code, len), len, debugstr_w(delimiter), flags, attr);
+
+    /* the code of a scriptlet is the body of a procedure, scanned the same way */
+    return get_text_attributes(code, len, delimiter, attr);
 }
 
 static HRESULT WINAPI VBScriptDebug_EnumCodeContextsOfPosition(IActiveScriptDebug *iface,
@@ -1049,11 +1553,136 @@ static HRESULT WINAPI VBScriptParse_AddScriptlet(IActiveScriptParse *iface,
         BSTR *pbstrName, EXCEPINFO *pexcepinfo)
 {
     VBScript *This = impl_from_IActiveScriptParse(iface);
-    FIXME("(%p)->(%s %s %s %s %s %s %s %lu %lx %p %p)\n", This, debugstr_w(pstrDefaultName),
+    named_item_t *item = NULL;
+    WCHAR *name = NULL, *src = NULL;
+    IDispatch *handler;
+    vbscode_t *code;
+    size_t len, code_len;
+    HRESULT hres;
+
+    TRACE("(%p)->(%s %s %s %s %s %s %s %lu %lx %p %p)\n", This, debugstr_w(pstrDefaultName),
           debugstr_w(pstrCode), debugstr_w(pstrItemName), debugstr_w(pstrSubItemName),
           debugstr_w(pstrEventName), debugstr_w(pstrDelimiter), wine_dbgstr_longlong(dwSourceContextCookie),
           ulStartingLineNumber, dwFlags, pbstrName, pexcepinfo);
-    return E_NOTIMPL;
+
+    if(!pbstrName || !pstrCode)
+        return E_POINTER;
+    *pbstrName = NULL;
+    if(pexcepinfo)
+        memset(pexcepinfo, 0, sizeof(*pexcepinfo));
+#ifdef SG_MUTANT_VBB1_SCRIPTLETEVT
+    pstrEventName = NULL;
+#endif
+
+    if(This->thread_id != GetCurrentThreadId() || This->state == SCRIPTSTATE_CLOSED || !This->is_initialized)
+        return E_UNEXPECTED;
+
+    if(pstrItemName) {
+        item = lookup_named_item(This->ctx, pstrItemName, 0);
+        if(!item) {
+            WARN("Unknown context %s\n", debugstr_w(pstrItemName));
+            return E_INVALIDARG;
+        }
+    }
+    if(!pstrDefaultName && !pstrEventName)
+        return E_INVALIDARG;
+
+    /* the scriptlet is named after the event it handles: Item_SubItem_Event */
+    if(pstrEventName && (pstrItemName || pstrSubItemName)) {
+        len = (pstrItemName ? wcslen(pstrItemName) + 1 : 0) + (pstrSubItemName ? wcslen(pstrSubItemName) + 1 : 0)
+              + wcslen(pstrEventName) + 1;
+        if(!(name = malloc(len * sizeof(WCHAR))))
+            return E_OUTOFMEMORY;
+        swprintf(name, len, L"%s%s%s%s%s", pstrItemName ? pstrItemName : L"", pstrItemName ? L"_" : L"",
+                 pstrSubItemName ? pstrSubItemName : L"", pstrSubItemName ? L"_" : L"", pstrEventName);
+    }else {
+        name = wcsdup(pstrDefaultName ? pstrDefaultName : pstrEventName);
+        if(!name)
+            return E_OUTOFMEMORY;
+    }
+
+    /* the code ends at the delimiter, if there is one */
+    code_len = wcslen(pstrCode);
+#ifndef SG_MUTANT_VBB1_SLETDELIM
+    if(pstrDelimiter && *pstrDelimiter) {
+        const WCHAR *end = wcsstr(pstrCode, pstrDelimiter);
+        if(end)
+            code_len = end - pstrCode;
+    }
+#endif
+
+    /* the code is the body of a procedure of that name without parameters */
+    len = wcslen(L"Sub ") + wcslen(name) + wcslen(L"()\r\n") + code_len + wcslen(L"\r\nEnd Sub") + 1;
+    if(!(src = malloc(len * sizeof(WCHAR)))) {
+        free(name);
+        return E_OUTOFMEMORY;
+    }
+    swprintf(src, len, L"Sub %s()\r\n%.*s\r\nEnd Sub", name, (int)code_len, pstrCode);
+
+    hres = compile_script(This->ctx, src, pstrItemName, NULL, dwSourceContextCookie,
+                          ulStartingLineNumber ? ulStartingLineNumber - 1 : 0,
+                          dwFlags & SCRIPTTEXT_ISPERSISTENT, &code);
+    free(src);
+    if(FAILED(hres)) {
+        free(name);
+        return hres;
+    }
+
+    hres = register_global_code(This->ctx, code, FALSE);
+    if(SUCCEEDED(hres) && pstrEventName && item && item->disp) {
+        IDispatch *target = item->disp;
+        IBindEventHandler *binder;
+
+        hres = create_func_ref(This->ctx, code->funcs, TRUE, &handler);
+        if(SUCCEEDED(hres)) {
+            if(pstrSubItemName) {
+                DISPPARAMS dp = {0};
+                VARIANT v;
+                DISPID id;
+                BSTR sub_name = SysAllocString(pstrSubItemName);
+
+                target = NULL;
+                hres = disp_get_id(item->disp, sub_name, VBDISP_CALLGET, FALSE, &id);
+                SysFreeString(sub_name);
+                if(SUCCEEDED(hres)) {
+                    V_VT(&v) = VT_EMPTY;
+                    hres = disp_call(This->ctx, item->disp, id, &dp, &v);
+                }
+                if(SUCCEEDED(hres)) {
+                    if(V_VT(&v) == VT_DISPATCH && V_DISPATCH(&v)) {
+                        target = V_DISPATCH(&v);
+                        IDispatch_AddRef(target);
+                    }else {
+                        hres = E_INVALIDARG;
+                    }
+                    VariantClear(&v);
+                }
+            }else {
+                IDispatch_AddRef(target);
+            }
+            if(SUCCEEDED(hres)) {
+                hres = IDispatch_QueryInterface(target, &IID_IBindEventHandler, (void**)&binder);
+                if(SUCCEEDED(hres)) {
+                    hres = IBindEventHandler_BindHandler(binder, pstrEventName, handler);
+                    IBindEventHandler_Release(binder);
+                }
+                IDispatch_Release(target);
+            }
+            IDispatch_Release(handler);
+        }
+    }
+
+    if(SUCCEEDED(hres)) {
+#ifdef SG_MUTANT_VBB1_SCRIPTLET
+        *pbstrName = SysAllocString(L"x");
+#else
+        *pbstrName = SysAllocString(name);
+#endif
+        if(!*pbstrName)
+            hres = E_OUTOFMEMORY;
+    }
+    free(name);
+    return hres;
 }
 
 static HRESULT WINAPI VBScriptParse_ParseScriptText(IActiveScriptParse *iface,
@@ -1257,6 +1886,8 @@ HRESULT WINAPI VBScriptFactory_CreateInstance(IClassFactory *iface, IUnknown *pU
     list_init(&ctx->code_list);
     list_init(&ctx->named_items);
     list_init(&ctx->func_refs);
+    list_init(&ctx->typelibs);
+    InitializeCriticalSection(&ctx->interrupt_cs);
 
     hres = init_global(ctx);
     if(FAILED(hres)) {
diff --git a/dlls/vbscript/vbscript.h b/dlls/vbscript/vbscript.h
index 90c9b5f..48338cf 100644
--- a/dlls/vbscript/vbscript.h
+++ b/dlls/vbscript/vbscript.h
@@ -217,6 +217,18 @@ struct _script_ctx_t {
      * of interp.c), the scope of Eval and Execute, and the GetRef objects */
     void *exec_top;
     struct list func_refs;
+
+    /* type libraries added with IActiveScript::AddTypeLib; their constants, once looked
+     * up, are kept (as constants of the global scope) in typelib_vars */
+    struct list typelibs;
+    struct _dynamic_var_t *typelib_vars;
+
+    /* IActiveScript::InterruptScriptThread, possibly from another thread */
+    CRITICAL_SECTION interrupt_cs;
+    LONG interrupt_pending;
+    DWORD interrupt_flags;
+    EXCEPINFO interrupt_ei;
+    BOOL aborting;   /* an interrupt without SCRIPTINTERRUPT_RAISEEXCEPTION unwinds all On Error handlers */
 };
 
 HRESULT init_global(script_ctx_t*);
@@ -408,6 +420,9 @@ typedef enum {
 #define SCRIPTTEXT_SG_NOREPORT 0x80000000  /* compile errors are raised, not reported to the site */
 HRESULT exec_dynamic_code(script_ctx_t*,const WCHAR*,dyncode_mode_t,VARIANT*);
 HRESULT register_global_code(script_ctx_t*,vbscode_t*,BOOL);
+VARIANT *lookup_typelib_const(script_ctx_t*,const WCHAR*);
+HRESULT create_func_ref(script_ctx_t*,function_t*,BOOL,IDispatch**);
+HRESULT take_interrupt(script_ctx_t*);
 void detach_func_refs(script_ctx_t*);
 void release_dynamic_var(dynamic_var_t*);
 named_item_t *lookup_named_item(script_ctx_t*,const WCHAR*,unsigned);
diff --git a/dlls/vbscript/vbscript.rc b/dlls/vbscript/vbscript.rc
index 805686d..2f2b749 100644
--- a/dlls/vbscript/vbscript.rc
+++ b/dlls/vbscript/vbscript.rc
@@ -44,6 +44,7 @@ STRINGTABLE
     VBSE_FOR_LOOP_NOT_INITIALIZED      "For loop not initialized"
     VBSE_ILLEGAL_NULL_USE              "Invalid use of Null"
     VBSE_CANT_CREATE_TMP_FILE          "Can't create necessary temporary file"
+    VBSE_OBJECT_REQUIRED               "Object required"
     VBSE_CANT_CREATE_OBJECT            "ActiveX component can't create object"
     VBSE_OLE_NOT_SUPPORTED             "Class doesn't support Automation"
     VBSE_OLE_FILE_NOT_FOUND            "File name or class name not found during Automation operation"
@@ -59,6 +60,7 @@ STRINGTABLE
     VBSE_INVALID_TYPELIB_VARIABLE      "Variable uses an Automation type not supported in VBScript"
     VBSE_SERVER_NOT_FOUND              "The remote server machine does not exist or is unavailable"
     VBSE_VARIABLE_UNDEFINED            "Variable is undefined"
+    VBSE_ILLEGAL_ASSIGNMENT            "Illegal assignment"
     VBSE_UNQUALIFIED_REFERENCE         "Invalid or unqualified reference"
     VBSE_SYNTAX_ERROR                  "Syntax error"
     VBSE_NAME_REDEFINED                "Name redefined"
diff --git a/dlls/vbscript/vbscript_defs.h b/dlls/vbscript/vbscript_defs.h
index 3feba1d..3c0de68 100644
--- a/dlls/vbscript/vbscript_defs.h
+++ b/dlls/vbscript/vbscript_defs.h
@@ -255,6 +255,7 @@
 #define VBSE_FOR_LOOP_NOT_INITIALIZED      92
 #define VBSE_ILLEGAL_NULL_USE              94
 #define VBSE_CANT_CREATE_TMP_FILE         322
+#define VBSE_OBJECT_REQUIRED               424
 #define VBSE_CANT_CREATE_OBJECT           429
 #define VBSE_OLE_NOT_SUPPORTED            430
 #define VBSE_OLE_FILE_NOT_FOUND           432
@@ -270,6 +271,7 @@
 #define VBSE_INVALID_TYPELIB_VARIABLE     458
 #define VBSE_SERVER_NOT_FOUND             462
 #define VBSE_VARIABLE_UNDEFINED           500
+#define VBSE_ILLEGAL_ASSIGNMENT            501
 #define VBSE_UNQUALIFIED_REFERENCE        505
 #define VBSE_SYNTAX_ERROR                1002
 #define VBSE_NAME_REDEFINED              1041
